<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Deconvolution of Bulk RNA-Seq Data Based on Deep Learning • digitalDLSorteR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Deconvolution of Bulk RNA-Seq Data Based on Deep Learning">
<meta property="og:description" content="Deconvolution of bulk RNA-Seq data using context-specific deconvolution models based on Deep Neural Networks using scRNA-Seq data as input. These models are able to make accurate estimates of the cell composition of bulk RNA-Seq samples from the same context using the advances provided by Deep Learning and the meaningful information provided by scRNA-Seq data. See Torroja and Sanchez-Cabo (2019) &lt;doi:10.3389/fgene.2019.00978&gt; for more details.">
<meta property="og:image" content="https://diegommcc.github.io/digitalDLSorteR/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">digitalDLSorteR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="./articles/pretrainedModels-hq.html">Using pre-trained context-specific deconvolution models</a>
    </li>
    <li>
      <a href="./articles/realModelWorkflow-hq.html">Building new deconvoluion models: deconvolution of colorectal cancer samples</a>
    </li>
    <li>
      <a href="./articles/kerasIssues.html">Keras/TensorFlow installation and configuration</a>
    </li>
    <li>
      <a href="./articles/hdf5Backend.html">HDF5 files as back-end</a>
    </li>
  </ul>
</li>
<li>
  <a href="reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/diegommcc/digitalDLSorteR" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="digitaldlsorter-">
<strong>digitalDLSorteR</strong> <img src="reference/figures/logo.png" align="right" width="120"><a class="anchor" aria-label="anchor" href="#digitaldlsorter-"></a>
</h1></div>
<p><a href="https://github.com/diegommcc/digitalDLsorteR/actions" class="external-link"><img src="https://github.com/diegommcc/digitalDLsorteR/workflows/R-CMD-check-bioc/badge.svg" alt="R build status"></a></p>
<div style="text-align:left">
<span>
<h4>
An R package to deconvolute bulk RNA-seq from scRNA-seq data based on Neural Networks
</h4>
<p></p></span>
</div>
<p><br></p>
<p>The <strong>digitalDLSorteR</strong> R package provides a set of tools to deconvolute cell type proportions of bulk RNA-seq data through the development of context-specific deconvolution models based on single-cell RNA-seq (scRNA-seq) data. These models are able to accurately estimate cell type proportions of bulk RNA-seq samples from specific biological environments. For more details about the algorithm and the functionalities implemented in this package, see <a href="https://www.frontiersin.org/articles/10.3389/fgene.2019.00978/full" class="external-link">Torroja and Sanchez-Cabo, 2019</a>, <a href="https://academic.oup.com/bioinformatics/article/40/2/btae072/7609037" class="external-link">Mañanes et al., 2024</a>, and <a href="https://diegommcc.github.io/digitalDLSorteR/" class="uri">https://diegommcc.github.io/digitalDLSorteR/</a>.</p>
<p><img src="reference/figures/summary.png"></p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><strong>digitalDLSorteR</strong> is available on CRAN and can be installed as follows:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"digitalDLSorteR"</span><span class="op">)</span></span></code></pre></div>
<p>The version under development is available on GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"remotes"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"diegommcc/digitalDLSorteR"</span><span class="op">)</span></span></code></pre></div>
<p>The package depends on the <a href="https://cran.r-project.org/package=tensorflow" class="external-link">tensorflow</a> R package, so a working Python interpreter with the Tensorflow Python library installed is needed. The <code>installTFpython</code> function provides an easy way to install a conda environment called <code>digitaldlsorter-env</code> with all necessary dependencies covered. We recommend installing the TensorFlow Python library in this way, although a custom installation is possible. See the <a href="https://diegommcc.github.io/digitalDLSorteR/articles/kerasIssues.html">Keras/TensorFlow installation and configuration</a> article of the package website for more details.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://diegommcc.github.io/digitalDLSorteR/">"digitalDLSorteR"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/installTFpython.html">installTFpython</a></span><span class="op">(</span>install.conda <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="rationale-of-digitaldlsorter">Rationale of <strong>digitalDLSorteR</strong>
<a class="anchor" aria-label="anchor" href="#rationale-of-digitaldlsorter"></a>
</h2>
<p>The algorithm consists of training Deep Neural Network (DNN) models with simulated bulk RNA-seq samples whose cell composition is known. These pseudo-bulk RNA-seq samples are generated by aggregating pre-characterized scRNA-seq data from specific biological environments. These models are able to accurately deconvolute new bulk RNA-seq samples from the same environment, as they are able to account for possible environmental-dependent transcriptional changes of specific cells, such as immune cells in complex diseases (e.g., specific subtypes of cancer or atherosclerosis). This aspect overcomes this limitation present in other methods. For instance, in the case of immune cells, published methods often rely on purified transcriptional profiles from peripheral blood mononuclear cells despite the fact that these cells are highly variable depending on environmental conditions. Thus, this feature together with the fact that scRNA-seq datasets improve over time (the more cells, the more variability learnt by the models) will lead to build more accurate and comprehensive models.</p>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>The package has two main ways of use:</p>
<ol style="list-style-type: decimal">
<li>Using pre-trained models included in the <strong>digitalLDSorteRmodels</strong> (<a href="https://github.com/diegommcc/digitalDLSorteRmodels" class="external-link uri">https://github.com/diegommcc/digitalDLSorteRmodels</a>) R package to deconvolute new bulk RNA-seq samples from the same environment. So far, the available models allow to deconvolute samples from human breast cancer (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75688" class="external-link">GSE75688</a> data from <a href="https://www.nature.com/articles/ncomms15081" class="external-link">Chung et al., 2017</a> used as reference), and colorectal cancer (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE132465" class="external-link">GSE132465</a>, <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE132257" class="external-link">GSE132257</a> and <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE144735" class="external-link">GSE144735</a> data from <a href="https://www.nature.com/articles/s41588-020-0636-z" class="external-link">Lee, Hong, Etlioglu Cho et al., 2020</a> used as reference). For more details about this workflow, please see the <a href="https://diegommcc.github.io/digitalDLSorteR/articles/pretrainedModels-hq.html">Using pre-trained context-specific deconvolution models</a> article. <strong>Disclaimer</strong>: these models intend to be a quick option to deconvolute samples from the same biological environment, but we strongly recommend generating new models with data manually curated by the users.</li>
<li>Building new deconvolution models from pre-characterized scRNA-seq datasets. This is the main way to use <strong>digitalDLSorteR</strong>. For more information on the workflow, see the article <a href="https://diegommcc.github.io/digitalDLSorteR/articles/realModelWorkflow-hq.html">Building new deconvolution models</a>.</li>
</ol>
<p>To use pre-trained context specific deconvolution models, <strong>digitalDLSorteR</strong> relies on the <strong>digitalDLSorteRmodels</strong> data R package. Therefore, it should be installed along with <strong>digitalDLSorteR</strong> from GitHub as follows:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"diegommcc/digitalDLSorteRmodels"</span><span class="op">)</span></span></code></pre></div>
<p>Once <strong>digitalDLSorteRmodels</strong> is loaded, the pre-trained models are available. See the article <a href="https://diegommcc.github.io/digitalDLSorteR/articles/pretrainedModels-hq.html">Using pre-trained context-specific deconvolution models</a> for an example.</p>
</div>
<div class="section level2">
<h2 id="final-remarks">Final remarks<a class="anchor" aria-label="anchor" href="#final-remarks"></a>
</h2>
<ul>
<li>Regarding pre-trained models, if you generate new models and want to make them available through the <strong>digitalDLSorteRmodels</strong> R package for other users to use them, contact with us!</li>
<li>We provide some pre-trained models that take into account genes that seem to be relevant for these environmental conditions. However, as these genes might be different depending on the bulk RNA-seq to be deconvoluted, we strongly recommend creating new models through the workflow explained <a href="https://diegommcc.github.io/digitalDLSorteR/articles/realModelWorkflow-hq.html">here</a>.</li>
<li>Contributions and suggestions are welcome!</li>
</ul>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>Please, if you use <strong>digitalDLSorteR</strong> in your research, cite <a href="https://www.frontiersin.org/articles/10.3389/fgene.2019.00978/full" class="external-link">Torroja and Sanchez-Cabo, 2019</a> (first description of the algorithm) and <a href="https://academic.oup.com/bioinformatics/article/40/2/btae072/7609037" class="external-link">Mañanes et al., 2024</a> (version for spatial transcriptomics data whose development has served to improve <strong>digitalDLSorteR</strong> as well).</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<table class="table">
<tr>
<td>
Chung, W., Eum, H. H., Lee, H. O., Lee, K. M., Lee, H. B., Kim, K. T., et al. (2017). Single-cell RNA-seq enables comprehensive tumour and immune cell profiling in primary breast cancer. <i>Nat. Commun.</i> <b>8</b> (1) 15081 <a href="https://doi.org/10.1038/ncomms15081" class="external-link">doi:10.1038/ncomms15081</a>
</td>
</tr>
<tr>
<td>
Lee, HO., Hong, Y., Etlioglu, H.E. et al. (2020). Lineage-dependent gene expression programs influence the immune landscape of colorectal cancer. <i>Nat. Genet.</i> <b>52</b> 594-603 <a href="https://doi.org/10.1038/s41588-020-0636-z" class="external-link">doi:10.1038/s41588-020-0636-z</a>
</td>
</tr>
<tr>
<td>
Torroja, C. and Sánchez-Cabo, F. (2019). digitalDLSorter: A Deep Learning algorithm to quantify immune cell populations based on scRNA-seq data. <i>Frontiers in Genetics</i> <b>10</b> 978 <a href="https://doi.org/10.3389/fgene.2019.00978" class="external-link">doi:10.3389/fgene.2019.00978</a>
</td>
</tr>
<tr>
<td>
Mañanes, D., Rivero-García, I., Relaño, C., Jimenez-Carretero, D., Torres, M., Sancho, D., Torroja, C. and Sánchez-Cabo, F. (2024). SpatialDDLS: An R package to deconvolute spatial transcriptomics data using neural networks. <i>Bioinformatics</i> <b>40</b> 2 <a href="https://doi.org/10.1093/bioinformatics/btae072" class="external-link">doi:10.1093/bioinformatics/btae072</a>
</td>
</tr>
</table>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=digitalDLSorteR" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/diegommcc/digitalDLSorteR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/diegommcc/digitalDLSorteR/issues" class="external-link">Report a bug</a></li>
<li><a href="https://github.com/diegommcc/digitalDLSorteRmodels" class="external-link">digitalDLSorteRmodels: pre-trained models for digitalDLSorteR</a></li>
<li><a href="https://github.com/diegommcc/digitalDLSorteRdata" class="external-link">digitalDLSorteRdata: example data for digitalDLSorteR</a></li>
<li><a href="https://diegommcc.github.io/SpatialDDLS/" class="external-link">SpatialDDLS: R package for deconvolution of spatial transcriptomics data</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing digitalDLSorteR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://github.com/diegommcc" class="external-link">Diego Mañanes</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0001-7247-6794" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://www.cnic.es/en/carlos-torroja" class="external-link">Carlos Torroja</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-8914-3400" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://www.cnic.es/en/fatima-sanchez-cabo" class="external-link">Fatima Sanchez-Cabo</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-1881-1664" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/diegommcc" class="external-link">Diego Mañanes</a>, <a href="https://www.cnic.es/en/carlos-torroja" class="external-link">Carlos Torroja</a>, <a href="https://www.cnic.es/en/fatima-sanchez-cabo" class="external-link">Fatima Sanchez-Cabo</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
