<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>digitalDLSorteR: an R package to deconvolute bulk RNA-Seq samples using single-cell RNA-seq data and neural networks — digitalDLSorteR • digitalDLSorteR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="digitalDLSorteR: an R package to deconvolute bulk RNA-Seq samples using single-cell RNA-seq data and neural networks — digitalDLSorteR"><meta property="og:description" content="digitalDLSorteR is an R package that allows to deconvolute bulk RNA-seq
data using context-specific deconvolution models based on single-cell RNA-seq
data and neural Networks. These models are able to make accurate
estimates of cell composition of bulk RNA-Seq samples from the same
context using the meaningful information provided by scRNA-seq data.
See Torroja and Sanchez-Cabo (2019) (doi:10.3389/fgene.2019.00978
) and
Mañanes et al., (2024) (doi:10.1093/bioinformatics/btae072
) for more
details."><meta property="og:image" content="https://diegommcc.github.io/digitalDLSorteR/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">digitalDLSorteR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>

  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href=".././articles/pretrainedModels-hq.html">Using pre-trained context-specific deconvolution models</a>
    </li>
    <li>
      <a href=".././articles/realModelWorkflow-hq.html">Building new deconvoluion models: deconvolution of colorectal cancer samples</a>
    </li>
    <li>
      <a href=".././articles/kerasIssues.html">Keras/TensorFlow installation and configuration</a>
    </li>
    <li>
      <a href=".././articles/hdf5Backend.html">HDF5 files as back-end</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/diegommcc/digitalDLSorteR" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>digitalDLSorteR: an R package to deconvolute bulk RNA-Seq samples using single-cell RNA-seq data and neural networks</h1>
    <small class="dont-index">Source: <a href="https://github.com/diegommcc/digitalDLSorteR/R/digitalDLSorteR.R" class="external-link"><code>R/digitalDLSorteR.R</code></a></small>
    <div class="hidden name"><code>digitalDLSorteR.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><span class="pkg">digitalDLSorteR</span> is an R package that allows to deconvolute bulk RNA-seq
data using context-specific deconvolution models based on single-cell RNA-seq
data and neural Networks. These models are able to make accurate
estimates of cell composition of bulk RNA-Seq samples from the same
context using the meaningful information provided by scRNA-seq data.
See Torroja and Sanchez-Cabo (2019) (<a href="https://doi.org/10.3389/fgene.2019.00978" class="external-link">doi:10.3389/fgene.2019.00978</a>
) and
Mañanes et al., (2024) (<a href="https://doi.org/10.1093/bioinformatics/btae072" class="external-link">doi:10.1093/bioinformatics/btae072</a>
) for more
details.</p>
    </div>


    <div id="details">
    <h2>Details</h2>
    <p>The method consists of a workflow that starts from
single-cell RNA-seq data and, after a few steps, a neural network model is
trained with simulated pseudo-bulk RNA-seq samples whose cell
composition is known. These trained models are able to deconvolute new bulk
RNA-seq samples from the same biological context. Its main advantage is the
possibility to build deconvolution models trained with real data from certain
biological environments. This fact tries to overcome the limitation
of other methods, since cell types may significantly change
their transcriptional profiles depending on tissue and disease context.</p>
<p>The package offers two usage ways: deconvolution of bulk RNA-seq samples
using pre-trained models available on the digitalDLSorteRmodels R package, or
building new deconvolution models from already identified scRNA-seq data.
See vignettes and <a href="https://diegommcc.github.io/digitalDLSorteR/">https://diegommcc.github.io/digitalDLSorteR/</a> for more
details.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>Useful links:</p><ul><li><p><a href="https://diegommcc.github.io/digitalDLSorteR/">https://diegommcc.github.io/digitalDLSorteR/</a></p></li>
<li><p><a href="https://github.com/diegommcc/digitalDLSorteR" class="external-link">https://github.com/diegommcc/digitalDLSorteR</a></p></li>
<li><p>Report bugs at <a href="https://github.com/diegommcc/digitalDLSorteR/issues" class="external-link">https://github.com/diegommcc/digitalDLSorteR/issues</a></p></li>
</ul></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p><strong>Maintainer</strong>: Diego Mañanes <a href="mailto:dmananesc@cnic.es">dmananesc@cnic.es</a> (<a href="https://orcid.org/0000-0001-7247-6794" class="external-link">ORCID</a>)</p>
<p>Authors:</p><ul><li><p>Carlos Torroja <a href="mailto:ctorroja@cnic.es">ctorroja@cnic.es</a> (<a href="https://orcid.org/0000-0001-8914-3400" class="external-link">ORCID</a>)</p></li>
<li><p>Fatima Sanchez-Cabo <a href="mailto:fscabo@cnic.es">fscabo@cnic.es</a> (<a href="https://orcid.org/0000-0003-1881-1664" class="external-link">ORCID</a>)</p></li>
</ul></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://github.com/diegommcc" class="external-link">Diego Mañanes</a>, <a href="https://www.cnic.es/en/carlos-torroja" class="external-link">Carlos Torroja</a>, <a href="https://www.cnic.es/en/fatima-sanchez-cabo" class="external-link">Fatima Sanchez-Cabo</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

