<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Estimate the parameters of the ZINB-WaVE model to simulate new single-cell
RNA-Seq expression profiles — estimateZinbwaveParams • digitalDLSorteR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Estimate the parameters of the ZINB-WaVE model to simulate new single-cell
RNA-Seq expression profiles — estimateZinbwaveParams"><meta property="og:description" content="Estimate the parameters of the ZINB-WaVE model using a real single-cell
RNA-Seq data set as reference to simulate new single-cell profiles and
increase the signal of underrepresented cell types. This step is optional,
only is needed if the size of you dataset is too small or there are
underrepresented cell types in order to train the Deep Neural Network model
in a more balanced way. After this step, the simSCProfiles
function will use the estimated parameters to simulate new single-cell
profiles. See ?simSCProfiles for more information."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">digitalDLSorteR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href=".././articles/pretrainedModels.html">Using pre-trained context-specific deconvolution models</a>
    </li>
    <li>
      <a href=".././articles/newModels.html">Building new deconvolution models</a>
    </li>
    <li>
      <a href=".././articles/realModelExample.html">Performance of a real model: deconvolution of colorectal cancer samples</a>
    </li>
    <li>
      <a href=".././articles/kerasIssues.html">Keras/TensorFlow installation and configuration</a>
    </li>
    <li>
      <a href=".././articles/hdf5Backend.html">HDF5 files as back-end</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/diegommcc/digitalDLSorteR" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Estimate the parameters of the ZINB-WaVE model to simulate new single-cell
RNA-Seq expression profiles</h1>
    <small class="dont-index">Source: <a href="https://github.com/diegommcc/digitalDLSorteR/R/simSingleCell.R" class="external-link"><code>R/simSingleCell.R</code></a></small>
    <div class="hidden name"><code>estimateZinbwaveParams.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Estimate the parameters of the ZINB-WaVE model using a real single-cell
RNA-Seq data set as reference to simulate new single-cell profiles and
increase the signal of underrepresented cell types. This step is optional,
only is needed if the size of you dataset is too small or there are
underrepresented cell types in order to train the Deep Neural Network model
in a more balanced way. After this step, the <code><a href="simSCProfiles.html">simSCProfiles</a></code>
function will use the estimated parameters to simulate new single-cell
profiles. See <code>?<a href="simSCProfiles.html">simSCProfiles</a></code> for more information.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">estimateZinbwaveParams</span><span class="op">(</span>
  <span class="va">object</span>,
  <span class="va">cell.type.column</span>,
  <span class="va">cell.ID.column</span>,
  <span class="va">gene.ID.column</span>,
  <span class="va">cell.cov.columns</span>,
  <span class="va">gene.cov.columns</span>,
  subset.cells <span class="op">=</span> <span class="cn">NULL</span>,
  proportional <span class="op">=</span> <span class="cn">TRUE</span>,
  set.type <span class="op">=</span> <span class="st">"All"</span>,
  threads <span class="op">=</span> <span class="fl">1</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p><code><a href="DigitalDLSorter-class.html">DigitalDLSorter</a></code> object with a
<code>single.cell.real</code> slot.</p></dd>
<dt>cell.type.column</dt>
<dd><p>Name or column number corresponding to the cell type
of each cell in cells metadata.</p></dd>
<dt>cell.ID.column</dt>
<dd><p>Name or column number corresponding to the cell names
of expression matrix in cells metadata.</p></dd>
<dt>gene.ID.column</dt>
<dd><p>Name or column number corresponding to the notation
used for features/genes in genes metadata.</p></dd>
<dt>cell.cov.columns</dt>
<dd><p>Name or column number(s) in cells metadata to be used
as covariates during model fitting (if no covariates are used, set to empty
or <code>NULL</code>).</p></dd>
<dt>gene.cov.columns</dt>
<dd><p>Name or column number(s) in genes metadata that will
be used as covariates during model fitting (if no covariates are used, set
to empty or <code>NULL</code>).</p></dd>
<dt>subset.cells</dt>
<dd><p>Number of cells to fit the ZINB-WaVE model. Useful when
the original data set is too large to fit the model. Set a value according
to the original data set and the resources available on your computer. If
<code>NULL</code> (by default), all cells will be used. Must be an integer
greater than or equal to the number of cell types considered and less than
or equal to the total number of cells.</p></dd>
<dt>proportional</dt>
<dd><p>If <code>TRUE</code>, the original cell type proportions in the
subset of cells generated by <code>subset.cells</code> will not be altered as far
as possible. If <code>FALSE</code>, all cell types will have the same number of
cells as far as possible (<code>TRUE</code> by default).</p></dd>
<dt>set.type</dt>
<dd><p>Cell type(s) to evaluate (<code>'All'</code> by default). It is
recommended fitting the model to all cell types rather than using only a
subset of them to capture the total variability present in the original
experiment even if only a subset of cell types is simulated.</p></dd>
<dt>threads</dt>
<dd><p>Number of threads used for estimation (1 by default). To set
up the parallel environment, the <span class="pkg">BiocParallel</span> package must be
installed.</p></dd>
<dt>verbose</dt>
<dd><p>Show informative messages during the execution (<code>TRUE</code> by
default).</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code><a href="DigitalDLSorter-class.html">DigitalDLSorter</a></code> object with <code>zinb.params</code>slot containing a <code>ZINBParams</code> object. This object
contains the estimated ZINB-WaVE parameters from the real single-cell
RNA-Se`q data.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>ZINB-WaVE is a flexible model for zero-inflated count data. This function
carries out the model fit to real single-cell data modeling \(Y_{ij}\) (the
count of feature \(j\) for sample \(i\)) as a random variable following a
zero-inflated negative binomial (ZINB) distribution. The estimated parameters
will be used for the simulation of new single-cell expression profiles by
sampling a negative binomial distribution and inserting dropouts from a
binomial distribution. To do so, <span class="pkg">digitalDLSorteR</span> uses the
<code><a href="https://rdrr.io/pkg/splatter/man/zinbEstimate.html" class="external-link">zinbEstimate</a></code> function from the <span class="pkg">splatter</span> package
(Zappia et al., 2017), a wrapper around the <code><a href="https://rdrr.io/pkg/zinbwave/man/zinbFit.html" class="external-link">zinbFit</a></code>
function from the <span class="pkg">zinbwave</span> package (Risso et al., 2018). For more
details about the model, see Risso et al., 2018.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Risso, D., Perraudeau, F., Gribkova, S. et al. (2018). A general
and flexible method for signal extraction from single-cell RNA-seq data.
Nat Commun 9, 284. doi: doi: <a href="https://doi.org/10.1038/s41467-017-02554-5" class="external-link">10.1038/s41467-017-02554-5</a>
.</p>
<p>Torroja, C. and Sánchez-Cabo, F. (2019). digitalDLSorter: A Deep Learning
algorithm to quantify immune cell populations based on scRNA-Seq data.
Frontiers in Genetics 10, 978. doi: doi: <a href="https://doi.org/10.3389/fgene.2019.00978" class="external-link">10.3389/fgene.2019.00978</a>
.</p>
<p>Zappia, L., Phipson, B. and Oshlack, A. Splatter: simulation of single-cell
RNA sequencing data. Genome Biol. 2017; 18: 174. doi:
doi: <a href="https://doi.org/10.1186/s13059-017-1305-0" class="external-link">10.1186/s13059-017-1305-0</a>
.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="simSCProfiles.html">simSCProfiles</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span> <span class="co"># reproducibility</span></span>
<span class="r-in"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">SingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span></span>
<span class="r-in">  assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span class="r-in">    counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span class="r-in">      <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">30</span>, lambda <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">15</span>, ncol <span class="op">=</span> <span class="fl">10</span>, </span>
<span class="r-in">      dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"RHC"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in">    <span class="op">)</span></span>
<span class="r-in">  <span class="op">)</span>,</span>
<span class="r-in">  colData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span class="r-in">    Cell_ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"RHC"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span>
<span class="r-in">    Cell_Type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"CellType"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">10</span>, </span>
<span class="r-in">                       replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span class="r-in">  <span class="op">)</span>,</span>
<span class="r-in">  rowData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span class="r-in">    Gene_ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in">  <span class="op">)</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"><span class="va">DDLS</span> <span class="op">&lt;-</span> <span class="fu"><a href="loadSCProfiles.html">loadSCProfiles</a></span><span class="op">(</span></span>
<span class="r-in">  single.cell.data <span class="op">=</span> <span class="va">sce</span>,</span>
<span class="r-in">  cell.ID.column <span class="op">=</span> <span class="st">"Cell_ID"</span>,</span>
<span class="r-in">  gene.ID.column <span class="op">=</span> <span class="st">"Gene_ID"</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-in"><span class="va">DDLS</span> <span class="op">&lt;-</span> <span class="fu">estimateZinbwaveParams</span><span class="op">(</span></span>
<span class="r-in">  object <span class="op">=</span> <span class="va">DDLS</span>,</span>
<span class="r-in">  cell.type.column <span class="op">=</span> <span class="st">"Cell_Type"</span>,</span>
<span class="r-in">  cell.ID.column <span class="op">=</span> <span class="st">"Cell_ID"</span>,</span>
<span class="r-in">  gene.ID.column <span class="op">=</span> <span class="st">"Gene_ID"</span>, </span>
<span class="r-in">  subset.cells <span class="op">=</span> <span class="fl">2</span>,</span>
<span class="r-in">  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Setting parallel environment to 1 thread(s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Estimating parameters for all cell types in the experiment</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Creating cell model matrix based on Cell_Type columns:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 	~Cell_Type</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Number of cells for each cell type:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - CellType1: 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - CellType2: 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Creating gene model matrix without gene covariates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Running estimation process (Start time 11:43:24 PM) </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing all zero genes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Fitting model...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Create model:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ok</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Initialize parameters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ok</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Optimize parameters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Iteration 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> penalized log-likelihood = -83.1203291378805</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> After dispersion optimization = -56.1335588623031</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    user  system elapsed </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.128   0.000   0.129 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> After right optimization = -54.6057641481274</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> After orthogonalization = -54.6057641481274</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    user  system elapsed </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.031   0.000   0.031 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> After left optimization = -54.6055360468269</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> After orthogonalization = -54.6055360468269</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Iteration 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> penalized log-likelihood = -54.6055360468269</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> After dispersion optimization = -54.6055360468269</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    user  system elapsed </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.114   0.000   0.115 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> After right optimization = -54.6055279954711</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> After orthogonalization = -54.6055279954711</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    user  system elapsed </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0.025   0.000   0.024 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> After left optimization = -54.6055197876521</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> After orthogonalization = -54.6055197876521</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Iteration 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> penalized log-likelihood = -54.6055197876521</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ok</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DONE</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Invested time: 3.92</span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://github.com/diegommcc" class="external-link">Diego Mañanes</a>, <a href="https://www.cnic.es/en/carlos-torroja" class="external-link">Carlos Torroja</a>, <a href="https://www.cnic.es/en/fatima-sanchez-cabo" class="external-link">Fatima Sanchez-Cabo</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

