<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Generate training and test cell composition matrices — generateBulkCellMatrix • digitalDLSorteR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Generate training and test cell composition matrices — generateBulkCellMatrix"><meta property="og:description" content="Generate training and test cell composition matrices for the simulation of
pseudo-bulk RNA-Seq samples with known cell composition using single-cell
expression profiles. The resulting ProbMatrixCellTypes
object contains a matrix that determines the proportion of the different cell
types that will compose the simulated pseudo-bulk samples. In addition, this
object also contains other information relevant for the process. This
function does not simulate pseudo-bulk samples, this task is performed by the
simBulkProfiles or trainDDLSModel
functions (see Documentation)."><meta property="og:image" content="https://diegommcc.github.io/digitalDLSorteR/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">digitalDLSorteR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>

  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href=".././articles/pretrainedModels-hq.html">Using pre-trained context-specific deconvolution models</a>
    </li>
    <li>
      <a href=".././articles/realModelWorkflow-hq.html">Building new deconvoluion models: deconvolution of colorectal cancer samples</a>
    </li>
    <li>
      <a href=".././articles/kerasIssues.html">Keras/TensorFlow installation and configuration</a>
    </li>
    <li>
      <a href=".././articles/hdf5Backend.html">HDF5 files as back-end</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/diegommcc/digitalDLSorteR" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate training and test cell composition matrices</h1>
    <small class="dont-index">Source: <a href="https://github.com/diegommcc/digitalDLSorteR/R/simBulk.R" class="external-link"><code>R/simBulk.R</code></a></small>
    <div class="hidden name"><code>generateBulkCellMatrix.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generate training and test cell composition matrices for the simulation of
pseudo-bulk RNA-Seq samples with known cell composition using single-cell
expression profiles. The resulting <code><a href="ProbMatrixCellTypes-class.html">ProbMatrixCellTypes</a></code>
object contains a matrix that determines the proportion of the different cell
types that will compose the simulated pseudo-bulk samples. In addition, this
object also contains other information relevant for the process. This
function does not simulate pseudo-bulk samples, this task is performed by the
<code><a href="simBulkProfiles.html">simBulkProfiles</a></code> or <code><a href="trainDDLSModel.html">trainDDLSModel</a></code>
functions (see Documentation).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">generateBulkCellMatrix</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">cell.ID.column</span>,</span>
<span>  <span class="va">cell.type.column</span>,</span>
<span>  <span class="va">prob.design</span>,</span>
<span>  <span class="va">num.bulk.samples</span>,</span>
<span>  n.cells <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  train.freq.cells <span class="op">=</span> <span class="fl">3</span><span class="op">/</span><span class="fl">4</span>,</span>
<span>  train.freq.bulk <span class="op">=</span> <span class="fl">3</span><span class="op">/</span><span class="fl">4</span>,</span>
<span>  proportion.method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">5</span>, <span class="fl">20</span>, <span class="fl">15</span>, <span class="fl">35</span>, <span class="fl">15</span><span class="op">)</span>,</span>
<span>  prob.sparsity <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  min.zero.prop <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  balanced.type.cells <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p><code><a href="DigitalDLSorter-class.html">DigitalDLSorter</a></code> object with
<code>single.cell.real</code> slot and, optionally, with <code>single.cell.simul</code>
slot.</p></dd>


<dt id="arg-cell-id-column">cell.ID.column<a class="anchor" aria-label="anchor" href="#arg-cell-id-column"></a></dt>
<dd><p>Name or column number corresponding to the cell names
of expression matrix in cells metadata.</p></dd>


<dt id="arg-cell-type-column">cell.type.column<a class="anchor" aria-label="anchor" href="#arg-cell-type-column"></a></dt>
<dd><p>Name or column number corresponding to the cell type
of each cell in cells metadata.</p></dd>


<dt id="arg-prob-design">prob.design<a class="anchor" aria-label="anchor" href="#arg-prob-design"></a></dt>
<dd><p>Data frame with the expected frequency ranges for each
cell type present in the experiment. This information can be estimated from
literature or from the single-cell experiment itself. This data frame must
be constructed by three columns with specific headings (see examples):</p><ul><li><p>A cell type column with the same name of the cell type
column in cells metadata (<code>cell.type.column</code>). If the name of the
column is not the same, the function will return an error. All cell types
must appear in the cells metadata.</p></li>
<li><p>A second column called
<code>'from'</code> with the start frequency for each cell type.</p></li>
<li><p>A third
column called <code>'to'</code> with the ending frequency for each cell type.</p></li>
</ul></dd>


<dt id="arg-num-bulk-samples">num.bulk.samples<a class="anchor" aria-label="anchor" href="#arg-num-bulk-samples"></a></dt>
<dd><p>Number of bulk RNA-Seq sample proportions (and thus
simulated bulk RNA-Seq samples) to be generated taking into account
training and test data. We recommend seting this value according to the
number of single-cell profiles available in
<code><a href="DigitalDLSorter-class.html">DigitalDLSorter</a></code> object avoiding an excesive
re-sampling, but generating a large number of samples for better training.</p></dd>


<dt id="arg-n-cells">n.cells<a class="anchor" aria-label="anchor" href="#arg-n-cells"></a></dt>
<dd><p>Number of cells that will be aggregated in order to simulate
one bulk RNA-Seq sample (100 by default).</p></dd>


<dt id="arg-train-freq-cells">train.freq.cells<a class="anchor" aria-label="anchor" href="#arg-train-freq-cells"></a></dt>
<dd><p>Proportion of cells used to simulate training
pseudo-bulk samples (2/3 by default).</p></dd>


<dt id="arg-train-freq-bulk">train.freq.bulk<a class="anchor" aria-label="anchor" href="#arg-train-freq-bulk"></a></dt>
<dd><p>Proportion of bulk RNA-Seq samples to the total number
(<code>num.bulk.samples</code>) used for the training set (2/3 by default).</p></dd>


<dt id="arg-proportion-method">proportion.method<a class="anchor" aria-label="anchor" href="#arg-proportion-method"></a></dt>
<dd><p>Vector of six integers that determines the
proportions of bulk samples generated by the different methods (see Details
and Torroja and Sanchez-Cabo, 2019. for more information). This vector
represents proportions, so its entries must add up 100. By default, a
majority of random samples will be generated without using predefined
ranges.</p></dd>


<dt id="arg-prob-sparsity">prob.sparsity<a class="anchor" aria-label="anchor" href="#arg-prob-sparsity"></a></dt>
<dd><p>It only affects the proportions generated by the first
method (Dirichlet distribution). It determines the probability of having
missing cell types in each simulated spot, as opposed to a mixture of all
cell types. A higher value for this parameter will result in more sparse
simulated samples.</p></dd>


<dt id="arg-min-zero-prop">min.zero.prop<a class="anchor" aria-label="anchor" href="#arg-min-zero-prop"></a></dt>
<dd><p>This parameter controls the minimum number of cell types
that will be absent in each simulated spot. If <code>NULL</code> (by default),
this value will be half of the total number of different cell types, but
increasing it will result in more spots composed of fewer cell types. This
helps to create more sparse proportions and cover a wider range of
situations during model training.</p></dd>


<dt id="arg-balanced-type-cells">balanced.type.cells<a class="anchor" aria-label="anchor" href="#arg-balanced-type-cells"></a></dt>
<dd><p>Boolean indicating whether the training and test
cells will be split in a balanced way considering the cell types
(<code>FALSE</code> by default).</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Show informative messages during the execution (<code>TRUE</code> by
default).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code><a href="DigitalDLSorter-class.html">DigitalDLSorter</a></code> object with
<code>prob.cell.types</code> slot containing a <code>list</code> with two
<code><a href="ProbMatrixCellTypes-class.html">ProbMatrixCellTypes</a></code> objects (training and test). For
more information about the structure of this class, see
<code>?<a href="ProbMatrixCellTypes-class.html">ProbMatrixCellTypes</a></code>.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>First, the available single-cell profiles are split into training and test
subsets (2/3 for training and 1/3 for test by default (see
<code>train.freq.cells</code>)) to avoid falsifying the results during model
evaluation. Next, <code>num.bulk.samples</code> bulk samples proportions are built
and the single-cell profiles to be used to simulate each pseudo-bulk RNA-Seq
sample are set, being 100 cells per bulk sample by default (see
<code>n.cells</code> argument). The proportions of training and test pseudo-bulk
samples are set by <code>train.freq.bulk</code> (2/3 for training and 1/3 for
testing by default). Finally, in order to avoid biases due to the composition
of the pseudo-bulk RNA-Seq samples, cell type proportions (\(w_1,...,w_k\),
where \(k\) is the number of cell types available in single-cell profiles)
are randomly generated by using six different approaches:</p>
<ol><li><p>Cell proportions are randomly sampled from a truncated
uniform distribution with predefined limits according to a priori knowledge
of the abundance of each cell type (see <code>prob.design</code> argument). This
information can be inferred from the single-cell experiment itself or from
the literature.</p></li>
<li><p>A second set is generated by randomly permuting cell
type labels from a distribution generated by the previous method.</p></li>
<li><p>Cell
proportions are randomly sampled as by method 1 without replacement.</p></li>
<li><p>Using the last method for generating proportions, cell types labels are
randomly sampled.</p></li>
<li><p>Cell proportions are randomly sampled from a
Dirichlet distribution.</p></li>
<li><p>Pseudo-bulk RNA-Seq samples composed of the
same cell type are generated in order to provide 'pure' pseudo-bulk samples.</p></li>
</ol><p>If you want to inspect the distribution of cell type proportions generated by
each method during the process, they can be visualized by the
<code><a href="showProbPlot.html">showProbPlot</a></code> function (see Documentation).</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Torroja, C. and Sánchez-Cabo, F. (2019). digitalDLSorter: A Deep
Learning algorithm to quantify immune cell populations based on scRNA-Seq
data. Frontiers in Genetics 10, 978. doi: <a href="https://doi.org/10.3389/fgene.2019.00978" class="external-link">doi:10.3389/fgene.2019.00978</a></p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="simBulkProfiles.html">simBulkProfiles</a></code>
<code><a href="ProbMatrixCellTypes-class.html">ProbMatrixCellTypes</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span> <span class="co"># reproducibility</span></span></span>
<span class="r-in"><span><span class="co"># simulated data</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">SingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>      <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">30</span>, lambda <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">15</span>, ncol <span class="op">=</span> <span class="fl">10</span>, </span></span>
<span class="r-in"><span>      dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"RHC"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span>,</span></span>
<span class="r-in"><span>  colData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    Cell_ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"RHC"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    Cell_Type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"CellType"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">10</span>, </span></span>
<span class="r-in"><span>                       replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span>,</span></span>
<span class="r-in"><span>  rowData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    Gene_ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">DDLS</span> <span class="op">&lt;-</span> <span class="fu"><a href="createDDLSobject.html">createDDLSobject</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  sc.data <span class="op">=</span> <span class="va">sce</span>,</span></span>
<span class="r-in"><span>  sc.cell.ID.column <span class="op">=</span> <span class="st">"Cell_ID"</span>,</span></span>
<span class="r-in"><span>  sc.gene.ID.column <span class="op">=</span> <span class="st">"Gene_ID"</span>,</span></span>
<span class="r-in"><span>  sc.filt.genes.cluster <span class="op">=</span> <span class="cn">FALSE</span>, </span></span>
<span class="r-in"><span>  sc.log.FC <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Bulk RNA-seq data not provided</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Processing single-cell data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>       - Filtering features:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>          - Selected features: 15</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>          - Discarded features: 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === No mitochondrial genes were found by using ^mt- as regrex</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Final number of dimensions for further analyses: 15</span>
<span class="r-in"><span><span class="va">probMatrixValid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  Cell_Type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"CellType"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">30</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">15</span>, <span class="fl">70</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">DDLS</span> <span class="op">&lt;-</span> <span class="fu">generateBulkCellMatrix</span><span class="op">(</span></span></span>
<span class="r-in"><span>  object <span class="op">=</span> <span class="va">DDLS</span>,</span></span>
<span class="r-in"><span>  cell.ID.column <span class="op">=</span> <span class="st">"Cell_ID"</span>,</span></span>
<span class="r-in"><span>  cell.type.column <span class="op">=</span> <span class="st">"Cell_Type"</span>,</span></span>
<span class="r-in"><span>  prob.design <span class="op">=</span> <span class="va">probMatrixValid</span>,</span></span>
<span class="r-in"><span>  num.bulk.samples <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === The number of bulk RNA-Seq samples that will be generated is equal to 10</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Training set cells by type:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - CellType1: 4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - CellType2: 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Test set cells by type:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - CellType1: 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - CellType2: 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Probability matrix for training data:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - Bulk RNA-Seq samples: 8</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - Cell types: 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Probability matrix for test data:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - Bulk RNA-Seq samples: 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - Cell types: 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DONE</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://github.com/diegommcc" class="external-link">Diego Mañanes</a>, <a href="https://www.cnic.es/en/carlos-torroja" class="external-link">Carlos Torroja</a>, <a href="https://www.cnic.es/en/fatima-sanchez-cabo" class="external-link">Fatima Sanchez-Cabo</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

