<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Calculate gradients of predicted cell types/loss function with respect to input features for interpreting trained deconvolution models — interGradientsDL • digitalDLSorteR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Calculate gradients of predicted cell types/loss function with respect to input features for interpreting trained deconvolution models — interGradientsDL"><meta property="og:description" content="This function enables users to gain insights into the interpretability of the
deconvolution model. It calculates the gradients of classes/loss function
with respect to the input features used in training. These numeric values are
calculated per gene and cell type in pure mixed transcriptional profiles,
providing information on the extent to which each feature influences the
model's prediction of cell proportions for each cell type."><meta property="og:image" content="https://diegommcc.github.io/digitalDLSorteR/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">digitalDLSorteR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href=".././articles/pretrainedModels-hq.html">Using pre-trained context-specific deconvolution models</a>
    </li>
    <li>
      <a href=".././articles/realModelWorkflow-hq.html">Building new deconvoluion models: deconvolution of colorectal cancer samples</a>
    </li>
    <li>
      <a href=".././articles/kerasIssues.html">Keras/TensorFlow installation and configuration</a>
    </li>
    <li>
      <a href=".././articles/hdf5Backend.html">HDF5 files as back-end</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/diegommcc/digitalDLSorteR" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Calculate gradients of predicted cell types/loss function with respect to input features for interpreting trained deconvolution models</h1>
    <small class="dont-index">Source: <a href="https://github.com/diegommcc/digitalDLSorteR/R/interGradientsDL.R" class="external-link"><code>R/interGradientsDL.R</code></a></small>
    <div class="hidden name"><code>interGradientsDL.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function enables users to gain insights into the interpretability of the
deconvolution model. It calculates the gradients of classes/loss function
with respect to the input features used in training. These numeric values are
calculated per gene and cell type in pure mixed transcriptional profiles,
providing information on the extent to which each feature influences the
model's prediction of cell proportions for each cell type.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">interGradientsDL</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"class"</span>,</span>
<span>  normalize <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scaling <span class="op">=</span> <span class="st">"standardize"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p><code><a href="DigitalDLSorter-class.html">DigitalDLSorter</a></code> object containing a trained
deconvolution model (<code>trained.model</code> slot) and pure mixed
transcriptional profiles (<code>bulk.simul</code> slot).</p></dd>


<dt>method</dt>
<dd><p>Method to calculate gradients with respect to inputs. It can be
<code>'class'</code> (gradients of predicted classes w.r.t. inputs),
<code>'loss'</code> (gradients of loss w.r.t. inputs) or <code>'both'</code>.</p></dd>


<dt>normalize</dt>
<dd><p>Whether to normalize data using logCPM (<code>TRUE</code> by
default). This parameter is only considered when the method used to
simulate the mixed transcriptional profiles (<code>simMixedProfiles</code>
function) was <code>"AddRawCount"</code>. Otherwise, data were already
normalized. This parameter should be set according to the transformation
used to train the model.</p></dd>


<dt>scaling</dt>
<dd><p>How to scale data. It can be: <code>"standardize"</code>
(values are centered around the mean with a unit standard deviation),
<code>"rescale"</code> (values are shifted and rescaled so that they end up
ranging between 0 and 1, by default) or <code>"none"</code> (no scaling is
performed). This parameter should be set according to the transformation
used to train the model.</p></dd>


<dt>verbose</dt>
<dd><p>Show informative messages during the execution (<code>TRUE</code> by
default).</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Object containing gradients in the <code>interpret.gradients</code> slot of
the <code>DigitalDLSorterDNN</code> object (<code>trained.model</code> slot).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Gradients of classes / loss function with respect to the input features are
calculated exclusively using pure mixed transcriptional profiles composed of
a single cell type. Consequently, these numbers can be interpreted as the
extent to which each feature is being used to predict each cell type
proportion. Gradients are calculated at the sample level for each gene, but
only mean gradients by cell type are reported. For additional details, see
Mañanes et al., 2024.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="deconvDDLSObj.html">deconvDDLSObj</a></code> <code><a href="plotTrainingHistory.html">plotTrainingHistory</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">SingleCellExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>      <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">30</span>, lambda <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">15</span>, ncol <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>      dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"RHC"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span>,</span></span>
<span class="r-in"><span>  colData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    Cell_ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"RHC"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    Cell_Type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"CellType"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>                       replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span>,</span></span>
<span class="r-in"><span>  rowData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    Gene_ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">DDLS</span> <span class="op">&lt;-</span> <span class="fu"><a href="createDDLSobject.html">createDDLSobject</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  sc.data <span class="op">=</span> <span class="va">sce</span>,</span></span>
<span class="r-in"><span>  sc.cell.ID.column <span class="op">=</span> <span class="st">"Cell_ID"</span>,</span></span>
<span class="r-in"><span>  sc.gene.ID.column <span class="op">=</span> <span class="st">"Gene_ID"</span>,</span></span>
<span class="r-in"><span>  sc.filt.genes.cluster <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Bulk RNA-seq data not provided</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Processing single-cell data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>       - Filtering features:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>          - Selected features: 15</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>          - Discarded features: 0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === No mitochondrial genes were found by using ^mt- as regrex</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Final number of dimensions for further analyses: 15</span>
<span class="r-in"><span><span class="va">prop.design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  Cell_Type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"CellType"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">30</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">15</span>, <span class="fl">70</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">DDLS</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateBulkCellMatrix.html">generateBulkCellMatrix</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  object <span class="op">=</span> <span class="va">DDLS</span>,</span></span>
<span class="r-in"><span>  cell.ID.column <span class="op">=</span> <span class="st">"Cell_ID"</span>,</span></span>
<span class="r-in"><span>  cell.type.column <span class="op">=</span> <span class="st">"Cell_Type"</span>,</span></span>
<span class="r-in"><span>  prob.design <span class="op">=</span> <span class="va">prop.design</span>, </span></span>
<span class="r-in"><span>  num.bulk.samples <span class="op">=</span> <span class="fl">50</span>,</span></span>
<span class="r-in"><span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === The number of bulk RNA-Seq samples that will be generated is equal to 50</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Training set cells by type:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - CellType1: 4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - CellType2: 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Test set cells by type:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - CellType1: 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - CellType2: 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Probability matrix for training data:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - Bulk RNA-Seq samples: 38</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - Cell types: 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Probability matrix for test data:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - Bulk RNA-Seq samples: 12</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     - Cell types: 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DONE</span>
<span class="r-in"><span><span class="va">DDLS</span> <span class="op">&lt;-</span> <span class="fu"><a href="simBulkProfiles.html">simBulkProfiles</a></span><span class="op">(</span><span class="va">DDLS</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Setting parallel environment to 1 thread(s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Generating train bulk samples:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Generating test bulk samples:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DONE</span>
<span class="r-in"><span><span class="va">DDLS</span> <span class="op">&lt;-</span> <span class="fu"><a href="trainDDLSModel.html">trainDDLSModel</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  object <span class="op">=</span> <span class="va">DDLS</span>,</span></span>
<span class="r-in"><span>  batch.size <span class="op">=</span> <span class="fl">12</span>,</span></span>
<span class="r-in"><span>  num.epochs <span class="op">=</span> <span class="fl">5</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Training and test from stored data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Using only simulated bulk samples</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     Using only simulated bulk samples</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Model: "DigitalDLSorter"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> _____________________________________________________________________</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Layer (type)                   Output Shape               Param #    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> =====================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Dense1 (Dense)                 (None, 200)                3200       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> _____________________________________________________________________</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BatchNormalization1 (BatchNorm (None, 200)                800        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> _____________________________________________________________________</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Activation1 (Activation)       (None, 200)                0          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> _____________________________________________________________________</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Dropout1 (Dropout)             (None, 200)                0          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> _____________________________________________________________________</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Dense2 (Dense)                 (None, 200)                40200      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> _____________________________________________________________________</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BatchNormalization2 (BatchNorm (None, 200)                800        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> _____________________________________________________________________</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Activation2 (Activation)       (None, 200)                0          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> _____________________________________________________________________</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Dropout2 (Dropout)             (None, 200)                0          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> _____________________________________________________________________</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Dense3 (Dense)                 (None, 2)                  402        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> _____________________________________________________________________</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> BatchNormalization3 (BatchNorm (None, 2)                  8          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> _____________________________________________________________________</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ActivationSoftmax (Activation) (None, 2)                  0          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> =====================================================================</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total params: 45,410</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Trainable params: 44,606</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Non-trainable params: 804</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> _____________________________________________________________________</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Training DNN with 38 samples:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Evaluating DNN in test data (12 samples)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>    - loss: NaN</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>    - accuracy: 0.1667</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>    - mean_absolute_error: NaN</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>    - categorical_accuracy: 0.1667</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> === Generating prediction results using test data</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> DONE</span>
<span class="r-in"><span><span class="co">## calculating gradients</span></span></span>
<span class="r-in"><span><span class="va">DDLS</span> <span class="op">&lt;-</span> <span class="fu">interGradientsDL</span><span class="op">(</span><span class="va">DDLS</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span>  </span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://github.com/diegommcc" class="external-link">Diego Mañanes</a>, <a href="https://www.cnic.es/en/carlos-torroja" class="external-link">Carlos Torroja</a>, <a href="https://www.cnic.es/en/fatima-sanchez-cabo" class="external-link">Fatima Sanchez-Cabo</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

