<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Building new deconvolution models • digitalDLSorteR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Building new deconvolution models">
<meta property="og:description" content="digitalDLSorteR">
<meta property="og:image" content="https://diegommcc.github.io/digitalDLSorteR/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">digitalDLSorteR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href=".././articles/pretrainedModels.html">Using pre-trained context-specific deconvolution models</a>
    </li>
    <li>
      <a href=".././articles/newModels.html">Building new deconvolution models</a>
    </li>
    <li>
      <a href=".././articles/realModelExample.html">Performance of a real model: deconvolution of colorectal cancer samples</a>
    </li>
    <li>
      <a href=".././articles/kerasIssues.html">Keras/TensorFlow installation and configuration</a>
    </li>
    <li>
      <a href=".././articles/hdf5Backend.html">HDF5 files as back-end</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/diegommcc/digitalDLSorteR" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Building new deconvolution models</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/diegommcc/digitalDLSorteR/vignettes/newModels.Rmd" class="external-link"><code>vignettes/newModels.Rmd</code></a></small>
      <div class="hidden name"><code>newModels.Rmd</code></div>

    </div>

    
    
<p><strong>digitalDLSorteR</strong> implements all the necessary tools
to build new context-specific deconvolution models from previously
identified single-cell RNA-Seq data. Taking into account the variability
of some cell types depending on the context in which they are found, we
aim to generate different models to each environment. From our view,
this opens a door to more accurate and specific models instead of using
generic transcriptional profiles as a reference (i.e., the use of
peripheral blood mononuclear cell (PBMC) to estimate the proportions of
tumor-infiltrating lymphocytes in oncological settings). In this
vignette, we will show the workflow required to build new
context-specific deconvolution models. For this, simulated data are
going to be used in order to avoid long runtimes. The performance of a
real model can be explored in the article <a href="realModelExample.html">Performance of a real model: deconvolution
of colorectal cancer samples</a>.</p>
<p>This workflow is computationally more expensive than using
pre-trained models, so we recommend building new models if you want to
deconvolute samples from an unavailable model or in case you think your
scRNA-Seq data provide a better picture of this environment than the
ones already offered in <code>digitalDLSorteRmodels</code>. In any case,
<strong>digitalDLSorteR</strong> provides a set of functionalities that
make this process easier and computationally cheaper in terms of RAM
usage: batch processing of data and the use of the <a href="https://bioconductor.org/packages/release/bioc/html/HDF5Array.html" class="external-link">HDF5Array</a>
and <a href="https://bioconductor.org/packages/release/bioc/html/DelayedArray.html" class="external-link">DelayedArray</a>
packages (see the article <a href="https://diegommcc.github.io/digitalDLSorteR/articles/hdf5Backend.html">HDF5
files as back-end</a> for more information). Furthermore, all steps are
centralized in the <code>DigitalDLSorter</code> S4-class, the core of
<strong>digitalDLSorteR</strong>, to provide a good user experience and
keep all the information tidy in the same object.</p>
<p>The main steps needed to build new models are summarized below, but
you can find a visual summary in the following figure.</p>
<ol style="list-style-type: decimal">
<li><a href="#load">Loading data into a <code>DigitalDLSorter</code>
object</a></li>
<li><a href="#oversample">Oversampling of single-cell profiles
(optional)</a></li>
<li><a href="#cell-prop">Generation of cell composition matrix for
pseudo-bulk RNA-Seq samples</a></li>
<li><a href="#simul-bulk">Simulation of pseudo-bulk RNA-Seq samples
using known cell composition</a></li>
<li><a href="#dnn">Deep Neural Network training</a></li>
<li><a href="#eval">Evaluation of trained deconvolution model on test
data: visualization of results</a></li>
<li><a href="#new-bulk">Loading and deconvolution of new bulk RNA-Seq
samples</a></li>
<li><a href="#save">Saving <code>DigitalDLSorter</code> object and
trained models</a></li>
</ol>
<!-- ```{r workflowfull_newModels, fig.cap = "Workflow to build new context-specific deconvolution models", echo = FALSE} --><!-- knitr::include_graphics("workflow_building_models.png") --><!-- ``` --><div class="section level3">
<h3 id="load">Loading data into a <code>DigitalDLSorter</code> object<a class="anchor" aria-label="anchor" href="#load"></a>
</h3>
<p>First, we have to load scRNA-Seq data into a
<code>DigitalDLSorter</code> object. This S4 class contains all the
slots needed to store the data generated during the construction of new
deconvolution models. The information needed consists of three
elements:</p>
<ul>
<li>Count matrix: a matrix with genes as rows and cells as columns.</li>
<li>Cells metadata: a table with annotations (columns) for each cell
(rows). The expected information in this data frame is a column with the
ID used for each cell, a column with the corresponding cell types, and
metadata that could be used as covariates in the following steps
(gender, sample type…).</li>
<li>Genes metadata with annotations (columns) for each gene (rows). This
data frame should contain the notation used for each gene in the counts
matrix and other covariates such as gene length, GC content, etc.</li>
</ul>
<p>This information may come from a pre-loaded
<code>SingleCellExperiment</code> object or from files stored on disk.
For the latter, tsv, tsv.gz, sparse matrices (mtx) and HDF5 (h5) formats
are accepted. Finally, data will be stored as a
<code>SingleCellExperiment</code> object in the
<code>single.cell.real</code> slot of the new
<code>DigitalDLSorter</code> object.</p>
<p>In addition, it is recommended to provide the bulk RNA-seq to be
deconvoluted at this point so that only genes actually relevant for the
deconvolution process are considered. The package will filter genes
according to different criteria so that only those important for the
task will be used for downstream steps.</p>
<p>In this tutorial, we will simulate both the scRNA-seq data used as a
reference and the bulk RNA-seq to be deconvoluted. This is done by
randomly sampling from a Poisson distribution, but importantly these
simulated data are only for descriptive purposes to show functionalities
of the package, they are not intended to generate realistic
transcriptomic data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## loading the packages</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://diegommcc.github.io/digitalDLSorteR/">"digitalDLSorteR"</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"SingleCellExperiment"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">"SummarizedExperiment"</a></span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>    <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">50000</span>, lambda <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">500</span>, ncol <span class="op">=</span> <span class="fl">100</span>, </span>
<span>    dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">500</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"RHC"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  colData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    Cell_ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"RHC"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    Cell_Type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span></span>
<span>      x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"CellType"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">100</span>, replace <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  rowData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    Gene_ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">500</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>    <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">10000</span>, lambda <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">500</span>, ncol <span class="op">=</span> <span class="fl">20</span>, </span>
<span>    dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">500</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sample_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  colData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    Sample_ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sample_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  rowData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    Gene_ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">500</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Then, we create the <code>DigitalDLSorter</code> object as
follows:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLSToy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createDDLSobject.html">createDDLSobject</a></span><span class="op">(</span></span>
<span>  sc.data <span class="op">=</span> <span class="va">sce</span>,</span>
<span>  sc.cell.ID.column <span class="op">=</span> <span class="st">"Cell_ID"</span>,</span>
<span>  sc.gene.ID.column <span class="op">=</span> <span class="st">"Gene_ID"</span>,</span>
<span>  sc.filt.genes.cluster <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  sc.log.FC <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  bulk.data <span class="op">=</span> <span class="va">se</span>,</span>
<span>  bulk.sample.ID.column <span class="op">=</span> <span class="st">"Sample_ID"</span>,</span>
<span>  bulk.gene.ID.column <span class="op">=</span> <span class="st">"Gene_ID"</span>,</span>
<span>  project <span class="op">=</span> <span class="st">"ToyExample"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">DDLSToy</span></span></code></pre></div>
<p>In documentation, you can see all the parameters that
<code>createDDLSobject</code> offers to process the loaded data, such as
<code>sc.min.counts</code> and <code>sc.min.cells</code>, etc. In this
case, we are setting <code>sc.log.FC = FALSE</code> due to the fact
these are simulated data and there is no biological signal in every cell
type. However, this parameter should be set to <code>TRUE</code> when
working with actual data.</p>
<p>In addition, in case of working with very large scRNA-Seq datasets,
<strong>digitalDLSorteR</strong> allows to use HDF5 files as back-end to
handle data that do not fit in RAM using the <a href="https://bioconductor.org/packages/release/bioc/html/HDF5Array.html" class="external-link">HDF5Array</a>
and <a href="https://bioconductor.org/packages/release/bioc/html/DelayedArray.html" class="external-link">DelayedArray</a>
packages. We only recommend them when analyzing huge amounts of data
that don’t fit in RAM. HDF5 files, despite being very powerful and
useful for dealing with RAM problems, make processes much slower. As an
example, the following code chunk would create an HDF5 file with the
scRNA-seq data that allows working without loading them into RAM. See
the documentation for more details.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLSToy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createDDLSobject.html">createDDLSobject</a></span><span class="op">(</span></span>
<span>sc.data <span class="op">=</span> <span class="va">sce</span>,</span>
<span>  sc.cell.ID.column <span class="op">=</span> <span class="st">"Cell_ID"</span>,</span>
<span>  sc.gene.ID.column <span class="op">=</span> <span class="st">"Gene_ID"</span>,</span>
<span>  sc.filt.genes.cluster <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  sc.log.FC <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  sc.gene.ID.column <span class="op">=</span> <span class="st">"external_gene_name"</span>,</span>
<span>  sc.file.backend <span class="op">=</span> <span class="st">"singlecell_data.h5"</span>,</span>
<span>  project <span class="op">=</span> <span class="st">"ToyExampleBreast"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="oversample">Oversampling of single-cell profiles<a class="anchor" aria-label="anchor" href="#oversample"></a>
</h3>
<p><strong>digitalDLSorteR</strong> offers the possibility to simulate
new single-cell profiles from real ones to increase signal and
variability in small datasets or when under-represented cell types are
present. This step is optional but recommended in these situations. The
<code>estimateZinbwaveParams</code> and <code>simSCProfiles</code>
functions are used for this purpose.</p>
<div class="section level4">
<h4 id="zinb">Tuning of the ZINB-WaVE model to simulate new single-cell
profiles<a class="anchor" aria-label="anchor" href="#zinb"></a>
</h4>
<p>First step is to estimate a set of parameters that fit the real
single-cell data to simulate new realistic single-cell profiles. We
chose the ZINB-WaVE framework <span class="citation">(Risso et al.
2018)</span> that estimates the parameters of a ZINB (zero-inflated
negative binomial) distribution. It was chosen for its ability to
accommodate not only variability within a particular cell type, but also
variability within the entire experiment.</p>
<p>This process is performed by the <code>estimateZinbwaveParams</code>
function, which makes use of the <a href="https://bioconductor.org/packages/release/bioc/html/zinbwave.html" class="external-link">zinbwave</a>
package. You must specify the column corresponding to cell types in the
cells metadata, and other cell/gene covariates can be added based on
your experimental design, such as patient, gender or gene length. This
process may take a few minutes to run, so be patient. In any case, you
can adjust the number of used threads in some steps during the
estimation with the <code>threads</code> argument depending on your
computational resources by the <a href="https://www.bioconductor.org/packages/release/bioc/html/BiocParallel.html" class="external-link">BiocParallel</a>
package. In case of large datasets with some cell types
under-represented, the <code>subset.cells</code> parameter allows making
a subset of cells to speed up the process. With the following code, a
total of 40 cells will be taken from the original scRNA-Seq data and
used to fit a ZINB-WaVE model.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLSToy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/estimateZinbwaveParams.html">estimateZinbwaveParams</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">DDLSToy</span>,</span>
<span>  cell.ID.column <span class="op">=</span> <span class="st">"Cell_ID"</span>,</span>
<span>  gene.ID.column <span class="op">=</span> <span class="st">"Gene_ID"</span>,</span>
<span>  cell.type.column <span class="op">=</span> <span class="st">"Cell_Type"</span>,</span>
<span>  subset.cells <span class="op">=</span> <span class="fl">40</span>,</span>
<span>  threads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLSToy</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="simul-sc">Simulating new single-cell profiles<a class="anchor" aria-label="anchor" href="#simul-sc"></a>
</h4>
<p>Once ZINB-WaVE parameters have been estimated, the
<code>simSCProfiles</code> function uses them to simulate new
single-cell profiles based on the real ones. It is done by randomly
sampling from a negative binomial distribution with the estimated ZINB
parameters <span class="math inline">\(\mu\)</span> and <span class="math inline">\(\theta\)</span>, and introducing dropouts by
sampling from a binomial distribution with the estimated probability
<span class="math inline">\(\pi\)</span>. You must specify the number of
cell profiles per cell type to be generated (<code>n.cells</code>). For
example, if your data set is composed of 5 cell types and
<code>n.cells</code> is equal to 10, the number of simulated profiles
will be 50.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLSToy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simSCProfiles.html">simSCProfiles</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">DDLSToy</span>,</span>
<span>  cell.ID.column <span class="op">=</span> <span class="st">"Cell_ID"</span>,</span>
<span>  cell.type.column <span class="op">=</span> <span class="st">"Cell_Type"</span>,</span>
<span>  n.cells <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  suffix.names <span class="op">=</span> <span class="st">"_Simul"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>These simulated single-cell profiles are stored in
<code>single.cell.simul</code> slot to be used to simulate new bulk
RNA-Seq profiles with a known cell composition.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLSToy</span></span></code></pre></div>
<p>In this step, it is also possible to store the new simulated
single-cell profiles in a HDF5 file. Indeed, they can be simulated in
batches, avoiding loading all data into RAM. The code would be as
follows:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLSToy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simSCProfiles.html">simSCProfiles</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">DDLSToy</span>,</span>
<span>  cell.ID.column <span class="op">=</span> <span class="st">"Cell_ID"</span>,</span>
<span>  cell.type.column <span class="op">=</span> <span class="st">"Cell_Type"</span>,</span>
<span>  n.cells <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  suffix.names <span class="op">=</span> <span class="st">"_Simul"</span>,</span>
<span>  file.backend <span class="op">=</span> <span class="st">"simulated_singlecell_data.h5"</span>,</span>
<span>  block.processing <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  block.size <span class="op">=</span> <span class="fl">20</span>, <span class="co"># number of single-cell profiles simulated per batch</span></span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="cell-prop">Generation of cell composition matrix for pseudo-bulk RNA-Seq
samples<a class="anchor" aria-label="anchor" href="#cell-prop"></a>
</h3>
<p>To simulate pseudobulk samples with a known cell composition, it is
necessary to generate a cell composition matrix that determines the
proportion of every cell type in every sample. This is carried out using
the <code>generateBulkCellMatrix</code> function that stores these
results in the <code>prob.cell.types</code> slot as a
<code>ProbMatrixCellTypes</code> object.</p>
<p>This process starts with dividing single-cell profiles into training
and test data (see <code>train.freq.cells</code> argument in
documentation). Each subset will be used to generate each subset of bulk
samples (training and test) in order to avoid any distortion of results
during model evaluation. Then, proportions are generated using six
different methods to avoid biases during training due to the cellular
composition of the simulated bulk RNA-Seq samples:</p>
<ol style="list-style-type: decimal">
<li>Cell proportions are randomly sampled from a truncated uniform
distribution with predefined limits according to <em>a priori</em>
knowledge of the abundance of each cell type (see
<code>prob.design</code> argument). This information can be inferred
from the single cell analysis itself or from the literature.</li>
<li>A second set is generated by randomly permuting cell type labels
from a distribution generated by the previous method.</li>
<li>Cell proportions are randomly sampled as by method 1 without
replacement.</li>
<li>Using the last method to generate proportions, cell types labels are
randomly sampled.</li>
<li>Cell proportions are randomly sampled from a Dirichlet
distribution.</li>
<li>Pseudo-bulk RNA-Seq samples composed of the same cell type are
generated in order to provide ‘pure’ pseudobulk samples.</li>
</ol>
<p>Proportion of each type of sample can be set by the
<code>proportion.method</code> argument. Moreover,
<code>prob.sparsity</code> controls the number of zeros (number of cell
types that will be zero in each sample) produced by each method. This
parameter was introduced in order to increase the level of sparsity in
the simulated cell composition matrices in order to increase the number
of situations the model is trained with. Finally, other important
parameters are <code>n.cells</code>, which determines the number of
cells that will compose each pseudobulk sample, and
<code>num.bulk.samples</code>, which defines the total number of
pseudobulk samples generated (training + test subsets). The code would
be as follows:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## prior knowledge for prob.design argument</span></span>
<span><span class="va">probMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  Cell_Type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"CellType"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, <span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">15</span>, <span class="fl">2</span><span class="op">)</span>, <span class="fl">50</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">70</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">DDLSToy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateBulkCellMatrix.html">generateBulkCellMatrix</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">DDLSToy</span>,</span>
<span>  cell.ID.column <span class="op">=</span> <span class="st">"Cell_ID"</span>,</span>
<span>  cell.type.column <span class="op">=</span> <span class="st">"Cell_Type"</span>,</span>
<span>  prob.design <span class="op">=</span> <span class="va">probMatrix</span>,</span>
<span>  num.bulk.samples <span class="op">=</span> <span class="fl">250</span>,</span>
<span>  n.cells <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLSToy</span></span></code></pre></div>
<p>Remember that this is a simulated example. In real circumstances,
depending on the number of single-cell profiles loaded/simulated at the
beginning and the computational resources, about 15,000-20,000 samples
would be recommended. You can inspect the cell composition matrix
created in this step with the getter function
<code>getProbMatrix</code>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/getProbMatrix.html">getProbMatrix</a></span><span class="op">(</span><span class="va">DDLSToy</span>, type.data <span class="op">=</span> <span class="st">"train"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="fu"><a href="../reference/getProbMatrix.html">getProbMatrix</a></span><span class="op">(</span><span class="va">DDLSToy</span>, type.data <span class="op">=</span> <span class="st">"train"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Moreover, distributions can be plotted using the
<code>showProbPlot</code> function:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span></span>
<span>  <span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/showProbPlot.html">showProbPlot</a></span><span class="op">(</span></span>
<span>      <span class="va">DDLSToy</span>, type.data <span class="op">=</span> <span class="st">"train"</span>, set <span class="op">=</span> <span class="va">x</span>, type.plot <span class="op">=</span> <span class="st">"boxplot"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="simul-bulk">Simulation of pseudo-bulk RNA-Seq samples with known cell
composition<a class="anchor" aria-label="anchor" href="#simul-bulk"></a>
</h3>
<p>Now, the simulated cell proportions are used to create the pseudobulk
samples. They are simulated by aggregating single-cell profiles of each
cell type according to these proportions. The idea is to simulate a real
bulk RNA-Seq data in which the gene expression levels of each cell are
aggregated into a single sample. Therefore, this expression matrix will
be generated according to the following equation:</p>
<p><span class="math display">\[\begin{equation}
  T_{ij} = \sum_{k = 1}^{K} \sum_{z = 1}^Z C_{izk}
\end{equation}\]</span></p>
<p><span class="math display">\[\begin{equation*}
  \textrm{such as} \left\{
\begin{array}{l}
  i = 1 \ldots M;\\
  j = 1 \ldots N \\
  Z = 1 \ldots \textrm{n.cells} \cdot P_{kj} \\
  \sum_{k = 1}^K Z \cdot P_{kj} = \textrm{n.cells}
\end{array}
\right.  
\end{equation*}\]</span></p>
<p>where <span class="math inline">\(T_{ij}\)</span> is the expression
level of gene <span class="math inline">\(i\)</span> in bulk sample
<span class="math inline">\(j\)</span>; <span class="math inline">\(C_{izk}\)</span> is the expression level of gene
<span class="math inline">\(i\)</span> in cell <span class="math inline">\(z\)</span> in bulk sample <span class="math inline">\(j\)</span>; and <span class="math inline">\(P_{kj}\)</span> is the proportion of cell type
<span class="math inline">\(k\)</span> in bulk sample <span class="math inline">\(j\)</span> (the cell composition matrix generated
in the previous step). <span class="math inline">\(Z\)</span> represents
the number of cells that will make up the proportion of cell type <span class="math inline">\(k\)</span> in the bulk sample <span class="math inline">\(j\)</span> and corresponds to the
<code>n.cells</code> parameter from the
<code>generateBulkCellMatrix</code> function. Cells are randomly sampled
based on their cell type and how they were divided into training and
test subsets. This step is performed by <code>simBulkProfiles</code> as
follows:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLSToy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simBulkProfiles.html">simBulkProfiles</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">DDLSToy</span>, type.data <span class="op">=</span> <span class="st">"both"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>These samples are stored as a <code>SummarizedExperiment</code>
object in the <code>bulk.simul</code> slot where they can be inspected
at any time:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLSToy</span></span></code></pre></div>
<p>The <code>simBulkProfiles</code> offers different ways to simulate
these pseudobulk samples, but in our experience, the best option is to
aggregate raw counts and normalize them afterwards
(<code>pseudobulk.function = "AddRawCount"</code>).</p>
<p>Again, these pseudobulk samples can be stored as an HDF5 file. This
is the most recommended step of <strong>digitalDLSorteR</strong> to use
this functionality, as it is the most computationally expensive part of
the package and these samples will only be accessed during training and
evaluation of Deep Neural Network (DNN) model. As in
<code>simSCProfiles</code>, samples can be simulated in batches and a
desired number of threads can also be set:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLSToy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simBulkProfiles.html">simBulkProfiles</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">DDLSToy</span>, </span>
<span>  type.data <span class="op">=</span> <span class="st">"both"</span>, </span>
<span>  file.backend <span class="op">=</span> <span class="st">"pseudobulk_samples.h5"</span>,</span>
<span>  block.processing <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  block.size <span class="op">=</span> <span class="fl">1000</span>, </span>
<span>  threads <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="deep-neural-network-training">Deep Neural Network training<a class="anchor" aria-label="anchor" href="#deep-neural-network-training"></a>
</h3>
<p>Once the pseudobulk samples have been generated, a deep neural
network can be trained and evaluated.
<code>trainDigitalDLSorterModel</code> is the function in charge of both
steps and uses the <a href="https://cran.r-project.org/package=keras" class="external-link">keras</a> package with
<a href="https://cran.r-project.org/package=tensorflow" class="external-link">tensorflow</a>
as back-end. If you want more information about <a href="https://cran.r-project.org/package=keras" class="external-link">keras</a> or have any
problems during its installation, please see <a href="https://diegommcc.github.io/digitalDLSorteR/articles/kerasIssues.html">Keras/TensorFlow
installation and configuration</a> for more details. In any case, the
<code>installTFpython</code> function automates this process, so we
recommend its use.</p>
<p>In terms of architecture and model parameters,
<code>trainDigitalDLSorterModel</code> implements by default two hidden
layers with 200 neurons each, although any of these parameters can be
modified through the <code>trainDigitalDLSorterModel</code> parameters.
In addition, for a more customized model, it is possible to provide a
pre-built model in the <code>custom.model</code> parameter. See the
documentation for more details.</p>
<p>The code with default parameters is as follows:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLSToy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trainDDLSModel.html">trainDDLSModel</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">DDLSToy</span>, scaling <span class="op">=</span> <span class="st">"standardize"</span>, batch.size <span class="op">=</span> <span class="fl">12</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><code>DDLSToy</code> will contain a <code>DigitalDLSorterDNN</code>
object with all the information associated with the model in the
<code>trained.model</code> slot: a
<code>keras.engine.sequential.Sequential</code> object with the trained
model, metrics and loss function histories during training, and
prediction results on test data.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLSToy</span></span></code></pre></div>
<p>Since this is a ‘toy’ example, results are not very accurate. For a
real example of a well trained model, see the <a href="realModelExample.html">Performance of a real model: deconvolution
of colorectal cancer samples</a> vignette.</p>
<div class="section level4">
<h4 id="on-the-fly-argument">
<code>on.the.fly</code> argument<a class="anchor" aria-label="anchor" href="#on-the-fly-argument"></a>
</h4>
<p>The <code>on.the.fly</code> argument of
<code>trainDigitalDLSorterModel</code> allows generating pseudobulk
samples ‘on the fly’. It means that it is possible to skip the
simulation of pseudobulk samples performed by the
<code>simBulkProfiles</code> function, and create the samples at the
same time as the neural network is being trained. Of course, running
times during training will increase, but data are not loaded into RAM or
stored in large HDF5 files. To use this functionality, it is only
necessary to set <code>on.the.fly = TRUE</code> as follows:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLSToy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trainDDLSModel.html">trainDDLSModel</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">DDLSToy</span>, on.the.fly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="eval">Evaluation of trained deconvolution model on test data:
visualization of results<a class="anchor" aria-label="anchor" href="#eval"></a>
</h3>
<p>While the prediction results on test data are informative about the
performance of the model, a more comprehensive analysis is needed. For
this task, <strong>digitalDLSorteR</strong> provides a set of
visualization functions to represent a variety of error metrics in
different ways.</p>
<p>First, <code>calculateEvalMetrics</code> is needed to calculate the
error metrics to be plotted. By default, absolute error
(<code>AbsErr</code>), proportional absolute error
(<code>ppAbsErr</code>), squared error (<code>SqrErr</code>) and
proportional squared error (<code>ppSqrErr</code>) are calculated for
every sample of test data. Furthermore, they are all aggregated using
their average values according to three criteria: each cell type
(<code>CellType</code>), proportion bins of 0.1 (<code>pBin</code>) and
number of different cell types (<code>nCellTypes</code>).</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLSToy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateEvalMetrics.html">calculateEvalMetrics</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">DDLSToy</span><span class="op">)</span></span></code></pre></div>
<p>Now, these results can be plotted by the following battery of
functions.</p>
<div class="section level4">
<h4 id="dist-err">
<code>distErrorPlot</code> and <code>barErrorPlot</code>: error
distributions<a class="anchor" aria-label="anchor" href="#dist-err"></a>
</h4>
<p>The <code>distErrorPlot</code> function allows plotting how errors
are distributed in different ways. Moreover, it allows to split the
plots in different panels representing how errors are distributed by a
given variable. Available variables are cell types
(<code>CellType</code>) and number of cell types in samples
(<code>nCellTypes</code>). In the following example, we will represent
the overall errors by cell type.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/distErrorPlot.html">distErrorPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">DDLSToy</span>,</span>
<span>  error <span class="op">=</span> <span class="st">"AbsErr"</span>,</span>
<span>  x.by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>  color.by <span class="op">=</span> <span class="st">"CellType"</span>, </span>
<span>  error.labels <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  type <span class="op">=</span> <span class="st">"boxplot"</span>,</span>
<span>  size.point <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Now, if you want to know if there is a bias towards a specific cell
type, yo can use <code>facet.by</code> parameter to split plots by cell
type:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/distErrorPlot.html">distErrorPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">DDLSToy</span>,</span>
<span>  error <span class="op">=</span> <span class="st">"AbsErr"</span>,</span>
<span>  facet.by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>  color.by <span class="op">=</span> <span class="st">"nCellTypes"</span>, </span>
<span>  type <span class="op">=</span> <span class="st">"violinplot"</span>,</span>
<span>  size.point <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>It is also possible to represent errors by number of different cell
types in samples:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/distErrorPlot.html">distErrorPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">DDLSToy</span>,</span>
<span>  error <span class="op">=</span> <span class="st">"AbsErr"</span>,</span>
<span>  color.by <span class="op">=</span> <span class="st">"CellType"</span>, </span>
<span>  facet.by <span class="op">=</span> <span class="st">"nCellTypes"</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"boxplot"</span>,</span>
<span>  size.point <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Finally, with <code>barErrorPlot</code>, the mean error values with
their corresponding dispersion ranges can be plotted as follows:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/barErrorPlot.html">barErrorPlot</a></span><span class="op">(</span><span class="va">DDLSToy</span>, error <span class="op">=</span> <span class="st">"MAE"</span>, by <span class="op">=</span> <span class="st">"CellType"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="corr-err">
<code>corrExpPredPlot</code>: correlation plots between predicted
and expected proportions<a class="anchor" aria-label="anchor" href="#corr-err"></a>
</h4>
<p>Ideally, the model should provide predictions that linearly match the
actual proportions. Therefore, you can generate correlation plots to
assess the model performance. By default, the Pearson’s coefficient
correlation (<span class="math inline">\(R\)</span>) and the concordance
correlation coefficient (<span class="math inline">\(CCC\)</span>) are
shown as annotations on the plots. The latter is a more realistic
measure as it decreases as the points move away from the identity.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/corrExpPredPlot.html">corrExpPredPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">DDLSToy</span>,</span>
<span>  color.by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>  size.point <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  corr <span class="op">=</span> <span class="st">"both"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>As in the previous case, plots can be split according to different
variables. Now, let’s split the results by <code>CellType</code> and
<code>nCellTypes</code> as an example:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/corrExpPredPlot.html">corrExpPredPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">DDLSToy</span>,</span>
<span>  color.by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>  facet.by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>  size.point <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  filter.sc <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  corr <span class="op">=</span> <span class="st">"both"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/corrExpPredPlot.html">corrExpPredPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">DDLSToy</span>,</span>
<span>  color.by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>  facet.by <span class="op">=</span> <span class="st">"nCellTypes"</span>,</span>
<span>  size.point <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  corr <span class="op">=</span> <span class="st">"both"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="bland-err">
<code>blandAltmanLehPlot</code>: Bland-Altman agreement plots<a class="anchor" aria-label="anchor" href="#bland-err"></a>
</h4>
<p>The <code>blandAltmanLehPlot</code> function allows to display
Bland-Altman agreement plots. This is a kind of graphical method for
comparing the level of agreement between two different sets of values.
The differences between predictions and actual proportions are plotted
against their averages. The central dashed line represents the mean
difference, while the two red dashed lines are the limits of agreement,
which are defined as the mean difference plus and minus 1.96 times the
standard deviation of the differences. 95% of the differences are
expected to fall between these two limits, so the wider the margins, the
worse the performance. It is also possible to display it in <span class="math inline">\(log_2\)</span> space.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/blandAltmanLehPlot.html">blandAltmanLehPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">DDLSToy</span>, </span>
<span>  color.by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>  log.2 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  size.point <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  filter.sc <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  density <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<p>In addition, this function has the same behaviour as previous ones,
as it is possible to split plots:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/blandAltmanLehPlot.html">blandAltmanLehPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">DDLSToy</span>, </span>
<span>  color.by <span class="op">=</span> <span class="st">"nCellTypes"</span>,</span>
<span>  facet.by <span class="op">=</span> <span class="st">"nCellTypes"</span>,</span>
<span>  log.2 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  size.point <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  filter.sc <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  density <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="new-bulk">Loading and deconvolution of new bulk RNA-Seq samples<a class="anchor" aria-label="anchor" href="#new-bulk"></a>
</h3>
<p>Once the model has been evaluated, we can deconvolute the bulk
RNA-seq data already loaded at the beginning of the pipeline. We could
load new data using the <code>loadDeconvData</code> function, but we
recommend doing it by using the <code>createDDLSobject</code> function,
since it will take only those genes shared between the two kinds of
data, so that only features relevant for the deconvolution process will
be used. Anyhow, to load new data, the call function would be as
follows:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">seExample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></span><span class="op">(</span>assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">countsBulk</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">DDLSToy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadDeconvData.html">loadDeconvData</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">DDLSToy</span>,</span>
<span>  data <span class="op">=</span> <span class="va">seExample</span>, </span>
<span>  name.data <span class="op">=</span> <span class="st">"Simulated.example"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>In our case, we can directly use the
<code>deconvDigitalDLSorterObj</code> function, which will deconvolute
the cell proportions of the cell types considered by the model. The
predicted proportions can be represented by the
<code>barPlotCellTypes</code> function. The cell composition matrix is
stored in the <code>deconv.results</code> slot.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLSToy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deconvDDLSObj.html">deconvDDLSObj</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">DDLSToy</span>, </span>
<span>  name.data <span class="op">=</span> <span class="st">"Bulk.DT"</span>,</span>
<span>  normalize <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  scaling <span class="op">=</span> <span class="st">"standardize"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## plot results</span></span>
<span><span class="fu"><a href="../reference/barPlotCellTypes.html">barPlotCellTypes</a></span><span class="op">(</span></span>
<span>  <span class="va">DDLSToy</span>, name.data <span class="op">=</span> <span class="st">"Bulk.DT"</span>, </span>
<span>  rm.x.text <span class="op">=</span> <span class="cn">TRUE</span>, color.line <span class="op">=</span> <span class="st">"black"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="save">Saving <code>DigitalDLSorter</code> object and trained models<a class="anchor" aria-label="anchor" href="#save"></a>
</h3>
<p><strong>digitalDLSorteR</strong> provides different ways to save
models on disk and to retrieve them in the <code>DigitalDLSorter</code>
object. First, you can save <code>DigitalDLSorter</code> objects as RDS
files. Since this file format only accepts native R objects, they are
not able to store complex data structures such as keras Python objects
(<code>keras.engine.sequential.Sequential</code> class). To make it
possible, <strong>digitalDLSorteR</strong> implements a
<code>saveRDS</code> generic function that converts the keras model
object into a list with network architecture and weights after training.
These two pieces of information are the minimal part needed to perform
new predictions. When the model is to be used, the model is compiled
back to a <code>keras.engine.sequential.Sequential</code> object.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## this code will not be run</span></span>
<span><span class="fu"><a href="../reference/saveRDS.html">saveRDS</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">DDLSToy</span>, file <span class="op">=</span> <span class="st">"valid/path"</span><span class="op">)</span></span></code></pre></div>
<p>However, the optimizer state is not saved in this way. To offer the
possibility to also save the optimizer, <strong>digitalDLSorteR</strong>
offers the <code>saveTrainedModelAsH5</code> function to save on disk
the whole neural network, and <code>loadTrainedModelFromH5</code> to
load-back models into <code>DigitalDLSorter</code> objects. Note that in
this way just the keras model is saved as an HDF5 file.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## this code will not be run</span></span>
<span><span class="fu"><a href="../reference/saveTrainedModelAsH5.html">saveTrainedModelAsH5</a></span><span class="op">(</span><span class="va">DDLSToy</span>, file.path <span class="op">=</span> <span class="st">"valid/path"</span><span class="op">)</span></span>
<span><span class="va">DDLSToy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadTrainedModelFromH5.html">loadTrainedModelFromH5</a></span><span class="op">(</span><span class="va">DDLSToy</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<!-- ## Session info {.unnumbered} -->
<!-- ```{r sessionInfo, echo=FALSE} -->
<!-- sessionInfo() -->
<!-- ``` -->
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Risso2018" class="csl-entry">
Risso, D., F. Perraudeau, S. Gribkova, S. Dudoit, and J. P. Vert. 2018.
<span>“<span class="nocase"><span>A</span> general and flexible method
for signal extraction from single-cell
<span>R</span><span>N</span><span>A</span>-seq data</span>.”</span>
<em>Nat Commun</em> 9 (1): 284.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/diegommcc" class="external-link">Diego Mañanes</a>, <a href="https://www.cnic.es/en/carlos-torroja" class="external-link">Carlos Torroja</a>, <a href="https://www.cnic.es/en/fatima-sanchez-cabo" class="external-link">Fatima Sanchez-Cabo</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
