<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using pre-trained context-specific deconvolution models • digitalDLSorteR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Using pre-trained context-specific deconvolution models">
<meta property="og:description" content="digitalDLSorteR">
<meta property="og:image" content="https://diegommcc.github.io/digitalDLSorteR/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">digitalDLSorteR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href=".././articles/pretrainedModels-hq.html">Using pre-trained context-specific deconvolution models</a>
    </li>
    <li>
      <a href=".././articles/realModelWorkflow-hq.html">Building new deconvoluion models: deconvolution of colorectal cancer samples</a>
    </li>
    <li>
      <a href=".././articles/kerasIssues.html">Keras/TensorFlow installation and configuration</a>
    </li>
    <li>
      <a href=".././articles/hdf5Backend.html">HDF5 files as back-end</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/diegommcc/digitalDLSorteR" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using pre-trained context-specific deconvolution
models</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/diegommcc/digitalDLSorteR/vignettes/pretrainedModels-hq.Rmd" class="external-link"><code>vignettes/pretrainedModels-hq.Rmd</code></a></small>
      <div class="hidden name"><code>pretrainedModels-hq.Rmd</code></div>

    </div>

    
    
<p><strong>digitalDLSorteR</strong> offers the possibility to use
pre-trained context-specific deconvolution models included in the
<strong>digitalDLSorteRmodels</strong> R package (<a href="https://github.com/diegommcc/digitalDLSorteRmodels" class="external-link uri">https://github.com/diegommcc/digitalDLSorteRmodels</a>) to
deconvolute new bulk RNA-seq samples from the same biological
environment. This is the simplest way to use
<strong>digitalDLSorteR</strong> and only requires loading into R a raw
bulk RNA-seq matrix with genes as rows (annotated as <em>SYMBOL</em>)
and samples as columns, and selecting the desired model. This is done by
the <code>deconvDDLSPretrained</code> function, which normalizes the new
samples to counts per million (CPMs) by default, so this matrix must be
provided as raw counts. Afterwards, estimated cell composition of each
sample can be explored as a bar chart using the
<code>barPlotCellTypes</code> function.</p>
<!-- ```{r workflow, fig.cap = "**Workflow to use pre-trained models**", echo = FALSE} -->
<!-- knitr::include_graphics("workflow_pretrained_models.png") -->
<!-- ``` -->
<!-- ![(#fig:workflow) Workflow to use pre-trained models](workflow_pretrained_models.png) -->
<div class="section level2">
<h2 id="available-models">Available models<a class="anchor" aria-label="anchor" href="#available-models"></a>
</h2>
<p>So far, available models only cover two possible biological
environments: breast cancer and colorectal cancer. These models are able
to accurately deconvolute new samples from the same environment as they
have been trained on.</p>
<div class="section level3">
<h3 id="breast-cancer-models">Breast cancer models<a class="anchor" aria-label="anchor" href="#breast-cancer-models"></a>
</h3>
<p>There are two deconvolution models for breast cancer samples that
differ in the level of specificity. Both have been trained using data
from <span class="citation">Chung et al. (2017)</span> (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75688" class="external-link">GSE75688</a>).</p>
<ul>
<li>
<code>breast.chung.generic</code>: it considers 13 cell types, four
of them being the intrinsic molecular subtypes of breast cancer
(<code>ER+</code>, <code>HER2+</code>, <code>ER+/HER2+</code> and
<code>TNBC</code>) and the rest immune and stromal cells
(<code>Stromal</code>, <code>Monocyte</code>, <code>TCD4mem</code>
(memory CD4+ T cells), <code>BGC</code> (germinal center B cells),
<code>Bmem</code> (memory B cells), <code>DC</code> (dendritic cells),
<code>Macrophage</code>, <code>TCD8</code> (CD8+ T cells) and
<code>TCD4reg</code> (regulatory CD4+ T cells)).</li>
<li>
<code>breast.chung.generic</code>: this model considers 7 cell types
that are generic groups of the cell types considered by the specific
version: B cells (<code>Bcell</code>), T CD4+ cells
(<code>TcellCD4</code>), T CD8+ cells (<code>TcellCD8</code>), monocytes
(<code>Monocyte</code>), dendritic cells (<code>DCs</code>), stromal
cells (<code>Stromal</code>) and tumor cells (<code>Tumor</code>).</li>
</ul>
</div>
<div class="section level3">
<h3 id="colorectal-cancer-model">Colorectal cancer model<a class="anchor" aria-label="anchor" href="#colorectal-cancer-model"></a>
</h3>
<p><code>DDLS.colon.lee</code> considers the following 22 cell types:
Anti-inflammatory_MFs (macrophages), B cells, CD4+ T cells, CD8+ T
cells, ECs (endothelial cells), ECs_tumor, Enterocytes, Epithelial
cells, Epithelial_cancer_cells, MFs_SPP1+, Mast cells, Myofibroblasts,
NK cells, Pericytes, Plasma_cells, Pro-inflammatory_MFs, Regulatory T
cells, Smooth muscle cells, Stromal cells, T follicular helper cells,
cDC (conventional dendritic cells), gamma delta T cells.</p>
<p>It has been generated using data from <span class="citation">Lee et
al. (2020)</span> (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE132465" class="external-link">GSE132465</a>,
<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE132257" class="external-link">GSE132257</a>
and <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE144735" class="external-link">GSE144735</a>).
The genes selected to train the model were defined by obtaining the
intersection between the scRNA-seq dataset and bulk RNA-seq data from
the The Cancer Genome Atlas (TCGA) project <span class="citation">(Koboldt et al. 2012; Ciriello et al. 2015)</span> and
using the <strong>digitalDLSorteR</strong>’s default parameters.</p>
</div>
</div>
<div class="section level2">
<h2 id="example-using-colorectal-samples-from-the-tcga-project">Example using colorectal samples from the TCGA project<a class="anchor" aria-label="anchor" href="#example-using-colorectal-samples-from-the-tcga-project"></a>
</h2>
<p>The following code chunk shows an example using the
<code>DDLS.colon.lee</code> model and data from TCGA loaded from
<code>digitalDLSorteRdata</code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://diegommcc.github.io/digitalDLSorteR/">"digitalDLSorteR"</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># to load pre-trained models</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"digitalDLSorteRmodels"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"diegommcc/digitalDLSorteRmodels"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/diegommcc/digitalDLSorteRmodels" class="external-link">digitalDLSorteRmodels</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># data for examples</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"digitalDLSorteRdata"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"diegommcc/digitalDLSorteRdata"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/diegommcc/digitalDLSorteRdata" class="external-link">"digitalDLSorteRdata"</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://dplyr.tidyverse.org" class="external-link">"dplyr"</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-data">Loading data<a class="anchor" aria-label="anchor" href="#loading-data"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># loading model from digitalDLSorteRmodel and example data from digitalDLSorteRdata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"DDLS.colon.lee"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"TCGA.colon.se"</span><span class="op">)</span></span></code></pre></div>
<p><code>DDLS.colon.lee</code> is a <code>DigitalDLSorterDNN</code>
object containing the trained model as well as specific information
about it, such as cell types considered, number of epochs used during
training, etc.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLS.colon.lee</span></span></code></pre></div>
<pre><code><span><span class="co">## Trained model: 60 epochs</span></span>
<span><span class="co">##   Training metrics (last epoch):</span></span>
<span><span class="co">##     loss: 0.113</span></span>
<span><span class="co">##     accuracy: 0.6851</span></span>
<span><span class="co">##     mean_absolute_error: 0.0131</span></span>
<span><span class="co">##     categorical_accuracy: 0.6851</span></span>
<span><span class="co">##   Evaluation metrics on test data:</span></span>
<span><span class="co">##     loss: 0.0979</span></span>
<span><span class="co">##     accuracy: 0.7353</span></span>
<span><span class="co">##     mean_absolute_error: 0.0117</span></span>
<span><span class="co">##     categorical_accuracy: 0.7353</span></span>
<span><span class="co">##   Performance evaluation over each sample: MAE MSE</span></span></code></pre>
<p>Here you can check the cell types considered by the model:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/cell.types.html">cell.types</a></span><span class="op">(</span><span class="va">DDLS.colon.lee</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span>collapse <span class="op">=</span> <span class="st">" / "</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Anti-inflammatory_MFs / B cells / CD4+ T cells / CD8+ T cells / ECs / ECs_tumor / Enterocytes / Epithelial cells / Epithelial_cancer_cells / MFs_SPP1+ / Mast cells / Myofibroblasts / NK cells / Pericytes / Plasma_cells / Pro-inflammatory_MFs / Regulatory T cells / Smooth muscle cells / Stromal cells / T follicular helper cells / cDC / gamma delta T cells"</span></span></code></pre>
<p>Now, we can use it to deconvolute <code>TCGA.colon.se</code> samples
as follows:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># deconvolution</span></span>
<span><span class="va">deconvResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deconvDDLSPretrained.html">deconvDDLSPretrained</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">TCGA.colon.se</span>,</span>
<span>  model <span class="op">=</span> <span class="va">DDLS.colon.lee</span>,</span>
<span>  normalize <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## === Filtering 57085 features in data that are not present in trained model</span></span></code></pre>
<pre><code><span><span class="co">## === Setting 0 features that are not present in trained model to zero</span></span></code></pre>
<pre><code><span><span class="co">## === Normalizing and scaling data</span></span></code></pre>
<pre><code><span><span class="co">## === Predicting cell types present in the provided samples</span></span></code></pre>
<pre><code><span><span class="co">##  1/17 [&gt;.............................] - ETA: 1s17/17 [==============================] - 0s 795us/step</span></span></code></pre>
<pre><code><span><span class="co">## DONE</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">deconvResults</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">deconvResults</span><span class="op">)</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">deconvResults</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          Anti-inflammatory_MFs     B cells CD4+ T cells CD8+ T cells</span></span>
<span><span class="co">## Sample_1            0.03555673 0.012629799 0.2033741921  0.009651527</span></span>
<span><span class="co">## Sample_2            0.04712304 0.066985354 0.0743350685  0.037740961</span></span>
<span><span class="co">## Sample_3            0.04327272 0.040529903 0.0704342425  0.065242790</span></span>
<span><span class="co">## Sample_4            0.06370532 0.087266177 0.0009350283  0.025520576</span></span>
<span><span class="co">## Sample_5            0.02191242 0.074413359 0.2854490578  0.011043761</span></span>
<span><span class="co">## Sample_6            0.05057831 0.009333177 0.0532203466  0.045282334</span></span>
<span><span class="co">##                  ECs  ECs_tumor Enterocytes Epithelial cells</span></span>
<span><span class="co">## Sample_1 0.005541964 0.01741249  0.00930732       0.07155290</span></span>
<span><span class="co">## Sample_2 0.034868974 0.04209921  0.04230683       0.03910230</span></span>
<span><span class="co">## Sample_3 0.005649934 0.05630094  0.04989067       0.05444700</span></span>
<span><span class="co">## Sample_4 0.063553527 0.06356231  0.04261205       0.01939960</span></span>
<span><span class="co">## Sample_5 0.006262527 0.06074687  0.04846477       0.01496454</span></span>
<span><span class="co">## Sample_6 0.007942557 0.05870197  0.01205215       0.15696421</span></span>
<span><span class="co">##          Epithelial_cancer_cells   MFs_SPP1+  Mast cells Myofibroblasts</span></span>
<span><span class="co">## Sample_1             0.059090309 0.005882894 0.037573170     0.01825435</span></span>
<span><span class="co">## Sample_2             0.027248615 0.064652205 0.043691449     0.06812528</span></span>
<span><span class="co">## Sample_3             0.044213723 0.028884979 0.046738215     0.05523874</span></span>
<span><span class="co">## Sample_4             0.062976733 0.021837069 0.006687692     0.02784960</span></span>
<span><span class="co">## Sample_5             0.030808710 0.039284099 0.021862393     0.02031602</span></span>
<span><span class="co">## Sample_6             0.005358715 0.020576829 0.106369697     0.02457042</span></span>
<span><span class="co">##            NK cells  Pericytes Plasma_cells Pro-inflammatory_MFs</span></span>
<span><span class="co">## Sample_1 0.20807564 0.04628118   0.03875882          0.004553901</span></span>
<span><span class="co">## Sample_2 0.07469995 0.02922468   0.01664688          0.024583302</span></span>
<span><span class="co">## Sample_3 0.05755382 0.02136339   0.04671244          0.007121266</span></span>
<span><span class="co">## Sample_4 0.06164910 0.04961385   0.03491507          0.019962544</span></span>
<span><span class="co">## Sample_5 0.04073286 0.02935413   0.02051810          0.061510801</span></span>
<span><span class="co">## Sample_6 0.03025281 0.01038809   0.04958449          0.001022621</span></span>
<span><span class="co">##          Regulatory T cells Smooth muscle cells Stromal cells</span></span>
<span><span class="co">## Sample_1       0.0008091909          0.00492954    0.02901827</span></span>
<span><span class="co">## Sample_2       0.0906273350          0.03798828    0.03085841</span></span>
<span><span class="co">## Sample_3       0.0678701103          0.03658660    0.08270273</span></span>
<span><span class="co">## Sample_4       0.0492653213          0.02225602    0.09293171</span></span>
<span><span class="co">## Sample_5       0.1044049561          0.02452374    0.04414547</span></span>
<span><span class="co">## Sample_6       0.1324384958          0.07235327    0.02527887</span></span>
<span><span class="co">##          T follicular helper cells         cDC gamma delta T cells</span></span>
<span><span class="co">## Sample_1               0.100938708 0.034195002          0.04661205</span></span>
<span><span class="co">## Sample_2               0.059172746 0.030084064          0.01783505</span></span>
<span><span class="co">## Sample_3               0.036124315 0.065181993          0.01793949</span></span>
<span><span class="co">## Sample_4               0.059893299 0.057660528          0.06594691</span></span>
<span><span class="co">## Sample_5               0.004997781 0.006953675          0.02732999</span></span>
<span><span class="co">## Sample_6               0.017107351 0.067269459          0.04335385</span></span></code></pre>
<p><code>deconvDDLSPretrained</code> returns a data frame with samples
as rows (<span class="math inline">\(k\)</span>) and cell types
considered by the model as columns (<span class="math inline">\(j\)</span>). Each entry corresponds to the
proportion of cell type <span class="math inline">\(k\)</span> in sample
<span class="math inline">\(i\)</span>. To visually evaluate these
results using a bar chart, you can use the <code>barplotCellTypes</code>
function as follows:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/barPlotCellTypes.html">barPlotCellTypes</a></span><span class="op">(</span></span>
<span>  <span class="va">deconvResults</span>, </span>
<span>  title <span class="op">=</span> <span class="st">"Results of deconvolution of TCGA colon samples"</span>, rm.x.text <span class="op">=</span> <span class="cn">T</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="resultsDeconvTCGA_pretrainedModels-1.png" width="2400" style="display: block; margin: auto;"></p>
<p>Let’s take 40 random samples just to improve the visualization:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/barPlotCellTypes.html">barPlotCellTypes</a></span><span class="op">(</span></span>
<span>  <span class="va">deconvResults</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">deconvResults</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">40</span><span class="op">)</span>, <span class="op">]</span>, </span>
<span>  title <span class="op">=</span> <span class="st">"Results of deconvolution of TCGA colon samples"</span>, rm.x.text <span class="op">=</span> <span class="cn">T</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="resultsDeconvTCGA_pretrainedModels_2-1.png" width="2400" style="display: block; margin: auto;"></p>
<p>Finally, <code>deconvDDLSPretrained</code> also offers two parameters
in case you want to simplify the results by aggregating cell proportions
of similar cell types: <code>simplify.set</code> and
<code>simplify.majority</code>. For instance, we can summarize different
CD4+ T cell subtypes into a unique label by using the
<code>simplify.set</code> parameter as follows:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># deconvolution</span></span>
<span><span class="va">deconvResultsSum</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deconvDDLSPretrained.html">deconvDDLSPretrained</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">TCGA.colon.se</span>,</span>
<span>  model <span class="op">=</span> <span class="va">DDLS.colon.lee</span>,</span>
<span>  normalize <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  simplify.set <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    `CD4+ T cells` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"CD4+ T cells"</span>, </span>
<span>      <span class="st">"T follicular helper cells"</span>, </span>
<span>      <span class="st">"gamma delta T cells"</span>, </span>
<span>      <span class="st">"Regulatory T cells"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## === Filtering 57085 features in data that are not present in trained model</span></span></code></pre>
<pre><code><span><span class="co">## === Setting 0 features that are not present in trained model to zero</span></span></code></pre>
<pre><code><span><span class="co">## === Normalizing and scaling data</span></span></code></pre>
<pre><code><span><span class="co">## === Predicting cell types present in the provided samples</span></span></code></pre>
<pre><code><span><span class="co">##  1/17 [&gt;.............................] - ETA: 0s17/17 [==============================] - 0s 844us/step</span></span></code></pre>
<pre><code><span><span class="co">## DONE</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">deconvResultsSum</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">deconvResults</span><span class="op">)</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/barPlotCellTypes.html">barPlotCellTypes</a></span><span class="op">(</span></span>
<span>  <span class="va">deconvResultsSum</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">deconvResultsSum</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">40</span><span class="op">)</span>, <span class="op">]</span>, </span>
<span>  title <span class="op">=</span> <span class="st">"Results of deconvolution of TCGA colon samples"</span>, rm.x.text <span class="op">=</span> <span class="cn">T</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="resultsDeconvTCGASimpl_pretrainedModels-1.png" width="1600" style="display: block; margin: auto;"></p>
<p>On the other hand, <code>simplify.majority</code> does not create new
classes but sums the proportions to the most abundant cell type from
those provided in each sample. See the documentation for more
details.</p>
</div>
<div class="section level2">
<h2 id="contribute-with-your-own-models">Contribute with your own models<a class="anchor" aria-label="anchor" href="#contribute-with-your-own-models"></a>
</h2>
<p>You can make available our own models to other users. Just drop an
email and we will make them available at the
<strong>digitalDLSorteRmodels</strong> R package!</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<!-- ## Session info {.unnumbered} -->
<!-- ```{r sessionInfo, echo=FALSE} -->
<!-- sessionInfo() -->
<!-- ``` -->
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Chung2017" class="csl-entry">
Chung, W., H. H. Eum, H. O. Lee, K. M. Lee, H. B. Lee, K. T. Kim, H. S.
Ryu, et al. 2017. <span>“<span class="nocase"><span>S</span>ingle-cell
<span>R</span><span>N</span><span>A</span>-seq enables comprehensive
tumour and immune cell profiling in primary breast
cancer</span>.”</span> <em>Nat Commun</em> 8: 15081.
</div>
<div id="ref-Ciriello2015" class="csl-entry">
Ciriello, G., M. L. Gatza, A. H. Beck, M. D. Wilkerson, S. K. Rhie, A.
Pastore, H. Zhang, et al. 2015. <span>“<span class="nocase"><span>C</span>omprehensive molecular portraits of
invasive lobular breast cancer</span>.”</span> <em>Cell</em> 163 (2):
506–19.
</div>
<div id="ref-Koboldt2012" class="csl-entry">
Koboldt, D. C., R. S. Fulton, M. D. McLellan, H. Schmidt, J.
Kalicki-Veizer, J. F. McMichael, L. L. Fulton, et al. 2012. <span>“<span class="nocase"><span>C</span>omprehensive molecular portraits of human
breast tumours</span>.”</span> <em>Nature</em> 490 (7418): 61–70.
</div>
<div id="ref-Lee2020" class="csl-entry">
Lee, H. O., Y. Hong, H. E. Etlioglu, Y. B. Cho, V. Pomella, B. Van den
Bosch, J. Vanhecke, et al. 2020. <span>“<span class="nocase"><span>L</span>ineage-dependent gene expression programs
influence the immune landscape of colorectal cancer</span>.”</span>
<em>Nat Genet</em> 52 (6): 594–603.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/diegommcc" class="external-link">Diego Mañanes</a>, <a href="https://www.cnic.es/en/carlos-torroja" class="external-link">Carlos Torroja</a>, <a href="https://www.cnic.es/en/fatima-sanchez-cabo" class="external-link">Fatima Sanchez-Cabo</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
