<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using pre-trained context-specific deconvolution models • digitalDLSorteR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Using pre-trained context-specific deconvolution models">
<meta property="og:description" content="digitalDLSorteR">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">digitalDLSorteR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href=".././articles/pretrainedModels.html">Using pre-trained context-specific deconvolution models</a>
    </li>
    <li>
      <a href=".././articles/newModels.html">Building new deconvolution models</a>
    </li>
    <li>
      <a href=".././articles/realModelExample.html">Performance of a real model: deconvolution of colorectal cancer samples</a>
    </li>
    <li>
      <a href=".././articles/kerasIssues.html">Keras/TensorFlow installation and configuration</a>
    </li>
    <li>
      <a href=".././articles/hdf5Backend.html">HDF5 files as back-end</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/diegommcc/digitalDLSorteR" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="pretrainedModels_files/header-attrs-2.10/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using pre-trained context-specific deconvolution models</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/diegommcc/digitalDLSorteR/vignettes/pretrainedModels.Rmd" class="external-link"><code>vignettes/pretrainedModels.Rmd</code></a></small>
      <div class="hidden name"><code>pretrainedModels.Rmd</code></div>

    </div>

    
    
<p><strong>digitalDLSorteR</strong> offers the possibility to use pre-trained context-specific deconvolution models included in <strong>digitalDLSorteRdata</strong> package (<a href="https://github.com/diegommcc/digitalDLSorteRdata" class="external-link uri">https://github.com/diegommcc/digitalDLSorteRdata</a>) to deconvolute new bulk RNA-Seq samples from the same biological environment. This is the simplest way to use <strong>digitalDLSorteR</strong> and only requires loading into R a raw bulk RNA-Seq matrix with genes as rows (annotated as <em>SYMBOL</em>) and samples as columns, and selecting the desired model. This is done by the <code>deconvDigitalDLSorter</code> function which normalizes the new samples to counts per million (CPMs) by default, so this matrix must be provided as raw counts. After the process, estimated cell composition of each sample can be explored as a bar chart using the <code>barPlotCellTypes</code> function (workflow summarized in the figure below).</p>
<div class="figure">
<img src="workflow_pretrained_models.png" alt="**Workflow to use pre-trained models**" width="811"><p class="caption">
<strong>Workflow to use pre-trained models</strong>
</p>
</div>
<!-- ![(#fig:workflow) Workflow to use pre-trained models](workflow_pretrained_models.png) -->
<div id="available-models" class="section level2">
<h2 class="hasAnchor">
<a href="#available-models" class="anchor" aria-hidden="true"></a>Available models</h2>
<p>So far, available models just cover two possible biological environments: breast cancer and colorectal cancer. These models are able to accurately deconvolute new samples from the same environment as they have been trained (with transcriptional profiles from these specific contexts).</p>
<div id="breast-cancer-models" class="section level3">
<h3 class="hasAnchor">
<a href="#breast-cancer-models" class="anchor" aria-hidden="true"></a>Breast cancer models</h3>
<p>There are two different deconvolution models for breast cancer samples that differ in the level of specificity. Both have been trained using data from <span class="citation">Chung et al. (2017)</span> (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75688" class="external-link">GSE75688</a>).</p>
<ul>
<li>
<code>breast.chung.generic</code>: this model considers 13 cell types, four of them being the intrinsic molecular subtypes of breast cancer (<code>ER+</code>, <code>HER2+</code>, <code>ER+/HER2+</code> and <code>TNBC</code>) and the rest immune and stromal cells (<code>Stromal</code>, <code>Monocyte</code>, <code>TCD4mem</code> (memory CD4+ T cells), <code>BGC</code> (germinal center B cells), <code>Bmem</code> (memory B cells), <code>DC</code> (dendritic cells), <code>Macrophage</code>, <code>TCD8</code> (CD8+ T cells) and <code>TCD4reg</code> (regulatory CD4+ T cells)).</li>
<li>
<code>breast.chung.generic</code>: this model considers 7 cell types, being generic groups of the cell types considered by the specific version: B cells (<code>Bcell</code>), T CD4+ cells (<code>TcellCD4</code>), T CD8+ cells (<code>TcellCD8</code>), monocytes (<code>Monocyte</code>), dendritic cells (<code>DCs</code>), stromal cells (<code>Stromal</code>) and tumor cells (<code>Tumor</code>).</li>
</ul>
</div>
<div id="colorectal-cancer-model" class="section level3">
<h3 class="hasAnchor">
<a href="#colorectal-cancer-model" class="anchor" aria-hidden="true"></a>Colorectal cancer model</h3>
<p>This model was created with data from <span class="citation">Li et al. (2017)</span> (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE81861" class="external-link">GSE81861</a>). It is called <code>colorectal.li</code> and considers 10 cell types: cancer cells (<code>CRC</code>), epithelial (<code>Ep</code>), monocytes (<code>M</code>), macrophages (<code>Mc</code>), fibroblasts (<code>Fb</code>), CD4 T cells (<code>CD4</code>), CD8 T cells (<code>CD8Gp</code>), CD8 T cells (<code>CD8Gn</code>), germinal center B cells (<code>gB</code>) and mature B cells (<code>pB</code>).</p>
</div>
</div>
<div id="example-using-breast-cancer-samples-from-tcga-project" class="section level2">
<h2 class="hasAnchor">
<a href="#example-using-breast-cancer-samples-from-tcga-project" class="anchor" aria-hidden="true"></a>Example using breast cancer samples from TCGA project</h2>
<p>The following code chunk shows an example using <code>breast.chung.generic</code> model and a small portion of data from The Cancer Genome Atlas (TCGA) project <span class="citation">(Koboldt et al. 2012; Ciriello et al. 2015)</span> loaded from <code>digitalDLSorteRdata</code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">digitalDLSorteR</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/diegommcc/digitalDLSorteRdata.git" class="external-link">digitalDLSorteRdata</a></span><span class="op">)</span><span class="op">)</span>
<span class="co"># disable eager execution for compatibility issues</span>
<span class="fu">tensorflow</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/tensorflow/man/tf.html" class="external-link">tf</a></span><span class="op">$</span><span class="va">compat</span><span class="op">$</span><span class="va">v1</span><span class="op">$</span><span class="fu">disable_eager_execution</span><span class="op">(</span><span class="op">)</span>
<span class="co"># loading model and example data from digitalDLSorteRdata</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">breast.chung.generic</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">TCGA.breast.small</span><span class="op">)</span></code></pre></div>
<p><code>breast.chung.generic</code> is a <code>DigitalDLSorterDNN</code> object that can be inspected to check which cell types are considered by the model, how well it performed on test data, etc.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">breast.chung.generic</span></code></pre></div>
<pre><code>## Trained model: 20 epochs
##   Training metrics (last epoch):
##     loss: 0.0222
##     accuracy: 0.9051
##     mean_absolute_error: 0.0186
##     categorical_accuracy: 0.9051
##   Evaluation metrics on test data:
##     loss: 0.0154
##     accuracy: 0.9629
##     mean_absolute_error: 0.0112
##     categorical_accuracy: 0.9629</code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/cell.types.html">cell.types</a></span><span class="op">(</span><span class="va">breast.chung.generic</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "Tumor"    "Stromal"  "Monocyte" "TcellCD4" "Bcell"    "DCs"      "TcellCD8"</code></pre>
<p>Now, we can use it to deconvolute <code>TCGA.breast.small</code> samples as follows:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># deconvolution</span>
<span class="va">deconvResults</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deconvDigitalDLSorter.html">deconvDigitalDLSorter</a></span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">TCGA.breast.small</span>,
  model <span class="op">=</span> <span class="va">breast.chung.generic</span>,
  normalize <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
<pre><code>## === Filtering 32876 features in data that are not present in trained model</code></pre>
<pre><code>## === Setting 11305 features that are not present in trained model to zero</code></pre>
<pre><code>## === Normalizing data</code></pre>
<pre><code>## === Predicting cell types present in the provided samples</code></pre>
<pre><code>## DONE</code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">deconvResults</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">deconvResults</span><span class="op">)</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">deconvResults</span><span class="op">)</span></code></pre></div>
<pre><code>##              Tumor     Stromal   Monocyte     TcellCD4        Bcell       DCs
## Sample_1 0.5716879 0.009549668 0.02705321 0.0009312486 2.085094e-05 0.2735237
## Sample_2 0.4737064 0.015726032 0.01532314 0.0008934744 1.881228e-05 0.3624099
## Sample_3 0.5632826 0.011810657 0.02118548 0.0009122468 2.213673e-05 0.3083328
## Sample_4 0.4798905 0.013469691 0.01989891 0.0007664470 4.999192e-05 0.3453779
## Sample_5 0.6283219 0.015287964 0.07622546 0.0022742241 3.257439e-06 0.1749346
## Sample_6 0.4960882 0.015261909 0.02924264 0.0010657794 1.409765e-05 0.3112175
##           TcellCD8
## Sample_1 0.1172334
## Sample_2 0.1319222
## Sample_3 0.0944540
## Sample_4 0.1405466
## Sample_5 0.1029526
## Sample_6 0.1471100</code></pre>
<p><code>deconvDigitalDLSorter</code> returns a data frame with the samples as rows (<span class="math inline">\(k\)</span>) and the cell types considered by the model as columns (<span class="math inline">\(j\)</span>). Each entry corresponds to the proportion of cell type <span class="math inline">\(k\)</span> in sample <span class="math inline">\(i\)</span>. To visually evaluate these results using a bar chart, you can use the <code>barplotCellTypes</code> function as follows:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/barPlotCellTypes.html">barPlotCellTypes</a></span><span class="op">(</span>
  <span class="va">deconvResults</span>, 
  color.line <span class="op">=</span> <span class="st">"black"</span>,
  title <span class="op">=</span> <span class="st">"Results of deconvolution of TCGA breast samples"</span>
<span class="op">)</span></code></pre></div>
<p><img src="pretrainedModels_files/figure-html/resultsDeconvTCGA-1.png" width="700"></p>
<p>Finally, <code>deconvDigitalDLSorterObj</code> also offers two parameters in case you want to simplify the results by aggregating cell proportions of similar cell types: <code>simplify.set</code> and <code>simplify.majority</code>. For instance, we can summarize dendritic cells (DCs) and monocytes in the new class mononuclear phagocyte system (MPS) using <code>simplify.set</code> as follows:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># deconvolution</span>
<span class="va">deconvResultsSum</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deconvDigitalDLSorter.html">deconvDigitalDLSorter</a></span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">TCGA.breast.small</span>,
  model <span class="op">=</span> <span class="va">breast.chung.generic</span>,
  normalize <span class="op">=</span> <span class="cn">TRUE</span>,
  simplify.set <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>MPS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DCs"</span>, <span class="st">"Monocyte"</span><span class="op">)</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<pre><code>## === Filtering 32876 features in data that are not present in trained model</code></pre>
<pre><code>## === Setting 11305 features that are not present in trained model to zero</code></pre>
<pre><code>## === Normalizing data</code></pre>
<pre><code>## === Predicting cell types present in the provided samples</code></pre>
<pre><code>## DONE</code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">deconvResultsSum</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">deconvResults</span><span class="op">)</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>
<span class="fu"><a href="../reference/barPlotCellTypes.html">barPlotCellTypes</a></span><span class="op">(</span>
  <span class="va">deconvResultsSum</span>, 
  rm.x.text <span class="op">=</span> <span class="cn">FALSE</span>,
  color.line <span class="op">=</span> <span class="st">"black"</span>,
  title <span class="op">=</span> <span class="st">"Results of deconvolution of TCGA breast samples (simplified)"</span>
<span class="op">)</span></code></pre></div>
<p><img src="pretrainedModels_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>On the other hand, <code>simplify.majority</code> does not create new classes but sums the proportions to the most abundant cell type from those provided in each sample. See the documentation and examples for more details.</p>
</div>
<div id="contribute-your-own-models" class="section level2">
<h2 class="hasAnchor">
<a href="#contribute-your-own-models" class="anchor" aria-hidden="true"></a>Contribute your own models</h2>
<p>New models are planned to be published soon, but you can contribute your own models by making them available to other users. Please do not hesitate to contact us! We’ll check the models and include them in the <strong>digitalDLSorteRdata</strong> package.</p>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor" aria-hidden="true"></a>References</h2>
<!-- ## Session info {.unnumbered} -->
<!-- ```{r sessionInfo, echo=FALSE} -->
<!-- sessionInfo() -->
<!-- ``` -->
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Chung2017" class="csl-entry">
Chung, W., H. H. Eum, H. O. Lee, K. M. Lee, H. B. Lee, K. T. Kim, H. S. Ryu, et al. 2017. <span>“<span class="nocase"><span>S</span>ingle-cell <span>R</span><span>N</span><span>A</span>-seq enables comprehensive tumour and immune cell profiling in primary breast cancer</span>.”</span> <em>Nat Commun</em> 8: 15081.
</div>
<div id="ref-Ciriello2015" class="csl-entry">
Ciriello, G., M. L. Gatza, A. H. Beck, M. D. Wilkerson, S. K. Rhie, A. Pastore, H. Zhang, et al. 2015. <span>“<span class="nocase"><span>C</span>omprehensive molecular portraits of invasive lobular breast cancer</span>.”</span> <em>Cell</em> 163 (2): 506–19.
</div>
<div id="ref-Koboldt2012" class="csl-entry">
Koboldt, D. C., R. S. Fulton, M. D. McLellan, H. Schmidt, J. Kalicki-Veizer, J. F. McMichael, L. L. Fulton, et al. 2012. <span>“<span class="nocase"><span>C</span>omprehensive molecular portraits of human breast tumours</span>.”</span> <em>Nature</em> 490 (7418): 61–70.
</div>
<div id="ref-Li2017" class="csl-entry">
Li, H., E. T. Courtois, D. Sengupta, Y. Tan, K. H. Chen, J. J. L. Goh, S. L. Kong, et al. 2017. <span>“<span class="nocase"><span>R</span>eference component analysis of single-cell transcriptomes elucidates cellular heterogeneity in human colorectal tumors</span>.”</span> <em>Nat Genet</em> 49 (5): 708–18.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/diegommcc" class="external-link external-link">Diego Mañanes</a>, <a href="https://www.cnic.es/en/carlos-torroja" class="external-link external-link">Carlos Torroja</a>, <a href="https://www.cnic.es/en/fatima-sanchez-cabo" class="external-link external-link">Fatima Sanchez-Cabo</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
