<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Building new deconvoluion models: deconvolution of colorectal cancer samples • digitalDLSorteR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Building new deconvoluion models: deconvolution of colorectal cancer samples">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">digitalDLSorteR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>

  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href=".././articles/pretrainedModels-hq.html">Using pre-trained context-specific deconvolution models</a>
    </li>
    <li>
      <a href=".././articles/realModelWorkflow-hq.html">Building new deconvoluion models: deconvolution of colorectal cancer samples</a>
    </li>
    <li>
      <a href=".././articles/kerasIssues.html">Keras/TensorFlow installation and configuration</a>
    </li>
    <li>
      <a href=".././articles/hdf5Backend.html">HDF5 files as back-end</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/diegommcc/digitalDLSorteR" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Building new deconvoluion models: deconvolution
of colorectal cancer samples</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/diegommcc/digitalDLSorteR/vignettes/realModelWorkflow-hq.Rmd" class="external-link"><code>vignettes/realModelWorkflow-hq.Rmd</code></a></small>
      <div class="hidden name"><code>realModelWorkflow-hq.Rmd</code></div>

    </div>

    
    
<p>In this example, we are going to reproduce the pre-trained model
<code>DDLS.colon.lee</code> available at the
<strong>digitalDLSorteRmodels</strong> R package. It was trained on data
from <span class="citation">Lee et al. (2020)</span> (<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE132465" class="external-link">GSE132465</a>,
<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE132257" class="external-link">GSE132257</a>
and <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE144735" class="external-link">GSE144735</a>),
and consist of ~ 100,000 cells from a total of 31 patients including
tumoral and healthy samples. These cells are divided into 22 cell types
covering the main ones found in this kind of samples:
<code>Anti-inflammatory_MFs</code> (macrophages), <code>B cells</code>,
<code>CD4+ T cells</code>, <code>CD8+ T cells</code>, <code>ECs</code>
(endothelial cells), <code>ECs_tumor</code>, <code>Enterocytes</code>,
<code>Epithelial cells</code>, <code>Epithelial_cancer_cells</code>,
<code>MFs_SPP1+</code>, <code>Mast cells</code>,
<code>Myofibroblasts</code>, <code>NK cells</code>,
<code>Pericytes</code>, <code>Plasma_cells</code>,
<code>Pro-inflammatory_MFs</code>, <code>Regulatory T cells</code>,
<code>Smooth muscle cells</code>, <code>Stromal cells</code>,
<code>T follicular helper cells</code>, <code>cDC</code> (conventional
dendritic cells), and <code>gamma delta T cells</code>. The expression
matrix only contains 2,000 genes selected by
<strong>digitalDLSorteR</strong> when the model was created to save time
and RAM. Thus, in this example we will set the parameters related to
gene filtering to zero.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">"SummarizedExperiment"</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"SingleCellExperiment"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://diegommcc.github.io/digitalDLSorteR/">"digitalDLSorteR"</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://dplyr.tidyverse.org" class="external-link">"dplyr"</a></span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"digitalDLSorteRdata"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"diegommcc/digitalDLSorteRdata"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/diegommcc/digitalDLSorteRdata" class="external-link">"digitalDLSorteRdata"</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://dplyr.tidyverse.org" class="external-link">"dplyr"</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://ggplot2.tidyverse.org" class="external-link">"ggplot2"</a></span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="loading-data">Loading data<a class="anchor" aria-label="anchor" href="#loading-data"></a>
</h2>
<p>We are also going to load bulk RNA-seq data on colorectal cancer
patients from the The Cancer Genome Atlas (TCGA) program <span class="citation">(Koboldt et al. 2012; Ciriello et al. 2015)</span>.
When building new deconvolution models, we recommend loading both the
single-cell RNA-seq reference and the bulk RNA-seq dataset to be
deconvoluted at the beginning so that <strong>digitalDLSorteR</strong>
can choose only genes actually relevant for the deconvolution
process.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"SCE.colon.Lee"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"TCGA.colon.se"</span><span class="op">)</span></span>
<span><span class="co"># to make it suitable for digitalDLSorteR</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">TCGA.colon.se</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">DataFrame</span><span class="op">(</span>SYMBOL <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">TCGA.colon.se</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><strong>Note:</strong> Please, note that here
<code>sc.filt.genes.cluster = FALSE</code> and
<code>sc.log.FC = FALSE</code> because we are providing with data
already preprocessed. However, we recommend keeping
<code>sc.filt.genes.cluster = TRUE</code> and
<code>sc.log.FC = TRUE</code> in regular situations, as
<strong>digitalDLSorteR</strong> will filter genes according to their
logFC.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLS.colon</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createDDLSobject.html">createDDLSobject</a></span><span class="op">(</span></span>
<span>  sc.data <span class="op">=</span> <span class="va">SCE.colon.Lee</span>,</span>
<span>  sc.cell.ID.column <span class="op">=</span> <span class="st">"Index"</span>,</span>
<span>  sc.gene.ID.column <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>  sc.cell.type.column <span class="op">=</span> <span class="st">"Cell_type_6"</span>,</span>
<span>  bulk.data <span class="op">=</span> <span class="va">TCGA.colon.se</span>,</span>
<span>  bulk.sample.ID.column <span class="op">=</span> <span class="st">"Bulk"</span>,</span>
<span>  <span class="va">bulk.gene.ID.column</span> <span class="op">&lt;-</span> <span class="st">"SYMBOL"</span>,</span>
<span>  filter.mt.genes <span class="op">=</span> <span class="st">"^MT-"</span>,</span>
<span>  sc.filt.genes.cluster <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  sc.log.FC <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  sc.min.counts <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  sc.min.cells <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  project <span class="op">=</span> <span class="st">"Colon-Cancer-Project"</span>  </span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## === Processing bulk transcriptomics data</span></span></code></pre>
<pre><code><span><span class="co">##       - Removing 2568 genes without expression in any cell</span></span></code></pre>
<pre><code><span><span class="co">## 'as(&lt;dgCMatrix&gt;, "dgTMatrix")' is deprecated.</span></span>
<span><span class="co">## Use 'as(., "TsparseMatrix")' instead.</span></span>
<span><span class="co">## See help("Deprecated") and help("Matrix-deprecated").</span></span></code></pre>
<pre><code><span><span class="co">##       - Filtering features:</span></span></code></pre>
<pre><code><span><span class="co">##          - Selected features: 54918</span></span></code></pre>
<pre><code><span><span class="co">##          - Discarded features: 1599</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## === Processing single-cell data</span></span></code></pre>
<pre><code><span><span class="co">##       - Filtering features:</span></span></code></pre>
<pre><code><span><span class="co">##          - Selected features: 2000</span></span></code></pre>
<pre><code><span><span class="co">##          - Discarded features: 0</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## === No mitochondrial genes were found by using ^MT- as regrex</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## === Final number of dimensions for further analyses: 2000</span></span></code></pre>
<p>After loading the data, we have a <code>DigitalDLSorter</code> object
with 2,000 genes and both the single-cell RNA-seq used as reference and
the bulk RNA-seq data to be deconvoluted.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLS.colon</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class DigitalDLSorter </span></span>
<span><span class="co">## Real single-cell profiles:</span></span>
<span><span class="co">##   2000 features and 106364 cells</span></span>
<span><span class="co">##   rownames: ADAMTS15 LY6G5B FBXO17 ... FBXO17 HLA-DPA1 ACKR3 CORIN </span></span>
<span><span class="co">##   colnames: SMC01-T_GGACAAGAGCTGGAAC SMC03-T_GCAAACTCAGCGATCC SMC05-N_TACTCATCAGATGAGC ... SMC05-N_TACTCATCAGATGAGC SMC06-T_TCTCTAAAGATCACGG SMC171-N-SING_GTATCTTAGTGGTAGC SMC13T-A1-F_CATTCGCGTCATGCAT </span></span>
<span><span class="co">## Bulk samples to deconvolute:</span></span>
<span><span class="co">##   Bulk.DT bulk samples:</span></span>
<span><span class="co">##     2000 features and 521 samples</span></span>
<span><span class="co">##     rownames: LINC00582 ACE2 MAL ... MAL MAGEA11 FZD10 COMP </span></span>
<span><span class="co">##     colnames: e79dd77e.6b51.4b49.ae8c.379c55ddae21 X9df77d76.ce4c.452d.9ce3.c0e1dc2360b8 X24c0b854.79c3.436a.87a7.4d0bbfac626a ... X24c0b854.79c3.436a.87a7.4d0bbfac626a dc4bba2d.8d49.4dcc.aa3c.17688fe73479 X8d445b92.e331.421c.9de0.d55a2fee908d X789c1767.4073.4ffa.b193.54b20b96d55e </span></span>
<span><span class="co">## Project: Colon-Cancer-Project</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="generating-cell-composition-matrix">Generating cell composition matrix<a class="anchor" aria-label="anchor" href="#generating-cell-composition-matrix"></a>
</h2>
<p>Now, let’s generate the cell composition matrix by using the
<code>generateBulkCellMatrix</code> function. It requires a data frame
with prior knowledge about how likely is to find each cell type in a
sample. For this example, we have used an approximation based on the
frequency of each cell type in each patient/sample from the scRNA-seq
dataset:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prop.design</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/single.cell.real.html">single.cell.real</a></span><span class="op">(</span><span class="va">DDLS.colon</span><span class="op">)</span><span class="op">@</span><span class="va">colData</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Patient</span>, <span class="va">Cell_type_6</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>Total <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Prop <span class="op">=</span> <span class="op">(</span><span class="va">Total</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">Total</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Cell_type_6</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>Prop_Mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">Prop</span><span class="op">)</span><span class="op">)</span>, Prop_SD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="va">Prop</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    from <span class="op">=</span> <span class="va">Prop_Mean</span>, </span>
<span>    to.1 <span class="op">=</span> <span class="va">Prop_Mean</span> <span class="op">*</span> <span class="op">(</span><span class="va">Prop_SD</span> <span class="op">*</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>    to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">to.1</span> <span class="op">&gt;</span> <span class="fl">100</span>, <span class="fl">100</span>, <span class="va">to.1</span><span class="op">)</span>,</span>
<span>    to.1 <span class="op">=</span> <span class="cn">NULL</span>, Prop_Mean <span class="op">=</span> <span class="cn">NULL</span>, Prop_SD <span class="op">=</span> <span class="cn">NULL</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `summarise()` has grouped output by 'Patient'. You can override using the</span></span>
<span><span class="co">## `.groups` argument.</span></span></code></pre>
<p>Then, we can generate the actual pseudobulk samples that will follow
these cell proportions. In this case, we generate 10,000 pseudobulk
samples (<code>num.bulk.samples</code>), although this number could be
increased according to available computational resources.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">DDLS.colon</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateBulkCellMatrix.html">generateBulkCellMatrix</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">DDLS.colon</span>,</span>
<span>  cell.ID.column <span class="op">=</span> <span class="st">"Index"</span>,</span>
<span>  cell.type.column <span class="op">=</span> <span class="st">"Cell_type_6"</span>,</span>
<span>  prob.design <span class="op">=</span> <span class="va">prop.design</span>,</span>
<span>  num.bulk.samples <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/simBulkProfiles.html">simBulkProfiles</a></span><span class="op">(</span>threads <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## === The number of bulk RNA-Seq samples that will be generated is equal to 10000</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## === Training set cells by type:</span></span></code></pre>
<pre><code><span><span class="co">##     - Anti-inflammatory_MFs: 501</span></span>
<span><span class="co">##     - B cells: 7336</span></span>
<span><span class="co">##     - CD4+ T cells: 10595</span></span>
<span><span class="co">##     - CD8+ T cells: 7570</span></span>
<span><span class="co">##     - cDC: 481</span></span>
<span><span class="co">##     - ECs: 874</span></span>
<span><span class="co">##     - ECs_tumor: 1398</span></span>
<span><span class="co">##     - Enterocytes: 920</span></span>
<span><span class="co">##     - Epithelial cells: 1411</span></span>
<span><span class="co">##     - Epithelial_cancer_cells: 19447</span></span>
<span><span class="co">##     - gamma delta T cells: 1459</span></span>
<span><span class="co">##     - Mast cells: 499</span></span>
<span><span class="co">##     - MFs_SPP1+: 4118</span></span>
<span><span class="co">##     - Myofibroblasts: 1789</span></span>
<span><span class="co">##     - NK cells: 1056</span></span>
<span><span class="co">##     - Pericytes: 677</span></span>
<span><span class="co">##     - Plasma_cells: 6259</span></span>
<span><span class="co">##     - Pro-inflammatory_MFs: 2342</span></span>
<span><span class="co">##     - Regulatory T cells: 4233</span></span>
<span><span class="co">##     - Smooth muscle cells: 614</span></span>
<span><span class="co">##     - Stromal cells: 5555</span></span>
<span><span class="co">##     - T follicular helper cells: 639</span></span></code></pre>
<pre><code><span><span class="co">## === Test set cells by type:</span></span></code></pre>
<pre><code><span><span class="co">##     - Anti-inflammatory_MFs: 155</span></span>
<span><span class="co">##     - B cells: 2456</span></span>
<span><span class="co">##     - CD4+ T cells: 3515</span></span>
<span><span class="co">##     - CD8+ T cells: 2583</span></span>
<span><span class="co">##     - cDC: 163</span></span>
<span><span class="co">##     - ECs: 336</span></span>
<span><span class="co">##     - ECs_tumor: 431</span></span>
<span><span class="co">##     - Enterocytes: 347</span></span>
<span><span class="co">##     - Epithelial cells: 439</span></span>
<span><span class="co">##     - Epithelial_cancer_cells: 6400</span></span>
<span><span class="co">##     - gamma delta T cells: 499</span></span>
<span><span class="co">##     - Mast cells: 175</span></span>
<span><span class="co">##     - MFs_SPP1+: 1406</span></span>
<span><span class="co">##     - Myofibroblasts: 592</span></span>
<span><span class="co">##     - NK cells: 325</span></span>
<span><span class="co">##     - Pericytes: 206</span></span>
<span><span class="co">##     - Plasma_cells: 2075</span></span>
<span><span class="co">##     - Pro-inflammatory_MFs: 761</span></span>
<span><span class="co">##     - Regulatory T cells: 1467</span></span>
<span><span class="co">##     - Smooth muscle cells: 191</span></span>
<span><span class="co">##     - Stromal cells: 1829</span></span>
<span><span class="co">##     - T follicular helper cells: 240</span></span></code></pre>
<pre><code><span><span class="co">## === Probability matrix for training data:</span></span></code></pre>
<pre><code><span><span class="co">##     - Bulk RNA-Seq samples: 7500</span></span>
<span><span class="co">##     - Cell types: 22</span></span></code></pre>
<pre><code><span><span class="co">## === Probability matrix for test data:</span></span></code></pre>
<pre><code><span><span class="co">##     - Bulk RNA-Seq samples: 2500</span></span>
<span><span class="co">##     - Cell types: 22</span></span></code></pre>
<pre><code><span><span class="co">## DONE</span></span></code></pre>
<pre><code><span><span class="co">## === Setting parallel environment to 2 thread(s)</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## === Generating train bulk samples:</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## === Generating test bulk samples:</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## DONE</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="neural-network-training">Neural network training<a class="anchor" aria-label="anchor" href="#neural-network-training"></a>
</h2>
<p>After generating the pseudobulk samples, we can train and evaluate
the model. The training step is only performed using cells/pseudobulk
samples coming from the training subset, since the test subset will be
used for the assessment of its performance.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLS.colon</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/trainDDLSModel.html">trainDDLSModel</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">DDLS.colon</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  1/79 [..............................] - ETA: 9s - loss: 0.1143 - accuracy: 0.7188 - mean_absolute_error: 0.0120 - categorical_accuracy: 0.718855/79 [===================&gt;..........] - ETA: 0s - loss: 0.1233 - accuracy: 0.6989 - mean_absolute_error: 0.0134 - categorical_accuracy: 0.698979/79 [==============================] - 0s 927us/step - loss: 0.1236 - accuracy: 0.7012 - mean_absolute_error: 0.0134 - categorical_accuracy: 0.7012</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="evaluation-of-the-model-on-test-data">Evaluation of the model on test data<a class="anchor" aria-label="anchor" href="#evaluation-of-the-model-on-test-data"></a>
</h2>
<p>Once the model is trained, we can explore how well the model behaves
on test samples. This step is critical because it allows us to assess if
<strong>digitalDLSorteR</strong> is actually understanding the signals
coming from each cell type or if on the contrary there are cell types
being ignored.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLS.colon</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateEvalMetrics.html">calculateEvalMetrics</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">DDLS.colon</span><span class="op">)</span></span></code></pre></div>
<p><strong>digitalDLSorteR</strong> implements different functions to
visualize the results and explore potential biases on the models. For
this tutorial, we will check the correlation between expected and
predicted proportions, but for a more detailed explanation about other
visualization functions, check the Documentation.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/corrExpPredPlot.html">corrExpPredPlot</a></span><span class="op">(</span></span>
<span>  <span class="va">DDLS.colon</span>,</span>
<span>  color.by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>  facet.by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>  corr <span class="op">=</span> <span class="st">"both"</span>, </span>
<span>  size.point <span class="op">=</span> <span class="fl">0.5</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre>
<p><img src="corr1_realModelWorkflow-1.png" width="2400" style="display: block; margin: auto;"></p>
<p>As it can be seen, the model is accurately predicting the cell
proportions of pseudobulk samples from the test data, which means that
it is detecting differential signals for each cell type.</p>
</div>
<div class="section level2">
<h2 id="deconvolution-of-tcga-samples">Deconvolution of TCGA samples<a class="anchor" aria-label="anchor" href="#deconvolution-of-tcga-samples"></a>
</h2>
<p>Now, to show its performance on real data, we are going to
deconvolute the samples from the TCGA project <span class="citation">(Koboldt et al. 2012; Ciriello et al. 2015)</span>
loaded at the beginning of the vignette. This dataset consists of 521
samples and includes both tumoral and healthy samples. This step is
performed by the <code>deconvDDLSObj</code> function, which will use the
trained model to obtain a set of predicted proportions for each sample
contained in the <code>deconv.data</code> slot.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLS.colon</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deconvDDLSObj.html">deconvDDLSObj</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">DDLS.colon</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    No 'name.data' provided. Using the first dataset</span></span></code></pre>
<p>We can plot the results as follows:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/barPlotCellTypes.html">barPlotCellTypes</a></span><span class="op">(</span><span class="va">DDLS.colon</span>, rm.x.text <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'name.data' not provided. By default, first results are used</span></span></code></pre>
<p><img src="barPlotResults_realModelWorkflow-1.png" width="2000" style="display: block; margin: auto;"></p>
<p>As the total number of samples is too high, we can see the results of
some samples by taking the predicted cell proportions and plotting 20
random samples with <code>barPlotCellTypes</code>:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span></span>
<span><span class="va">resDeconvTCGA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/deconv.results.html">deconv.results</a></span><span class="op">(</span><span class="va">DDLS.colon</span>, name.data <span class="op">=</span> <span class="st">"Bulk.DT"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/barPlotCellTypes.html">barPlotCellTypes</a></span><span class="op">(</span></span>
<span>  <span class="va">resDeconvTCGA</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">521</span>, size <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>, <span class="op">]</span>, rm.x.text <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"Results of deconvolution (20 random samples)"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="barPlotResults20_realModelWorkflow-1.png" width="2000" style="display: block; margin: auto;"></p>
<p>Now, we can represent the cell proportions of every cell type
considered by the model separating healthy and tumoral samples. We are
also going to filter out samples considered metastatic or recurrent
(check the <code>TCGA.colon.se</code> object) because these groups are
composed of only 1 sample:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  Sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">resDeconvTCGA</span><span class="op">)</span>,</span>
<span>  TypeSample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">TCGA.colon.se</span><span class="op">)</span><span class="op">[[</span><span class="st">"Tumor_Type"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">resDeconvTCGA</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html" class="external-link">melt</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">TypeSample</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Metastatic"</span>, <span class="st">"Recurrent"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">TypeSample</span>, y <span class="op">=</span> <span class="va">value</span>, fill <span class="op">=</span> <span class="va">variable</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">variable</span>, scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">digitalDLSorteR</span><span class="fu">:::</span><span class="fu">default.colors</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Estimated proportions in TCGA data (all cell types)"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>      plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span>, hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>      legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Using Sample, TypeSample as id variables</span></span></code></pre>
<p><img src="boxplotResults_realModelWorkflow-1.png" width="3600" style="display: block; margin: auto;"></p>
<p>In general, the results seem to be in line with what it is known:
tumoral samples show a huge immune infiltration, whereas other cell
types such as epithelial cells are displaced. We can also specifically
inspect the predicted proportions of enterocytes, tumor, epithelial, and
stromal cells:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  Sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">resDeconvTCGA</span><span class="op">)</span>,</span>
<span>  CRC <span class="op">=</span> <span class="va">resDeconvTCGA</span><span class="op">[</span>, <span class="st">"Epithelial_cancer_cells"</span><span class="op">]</span>,</span>
<span>  Epithelial <span class="op">=</span> <span class="va">resDeconvTCGA</span><span class="op">[</span>, <span class="st">"Epithelial cells"</span><span class="op">]</span>,</span>
<span>  Stromal <span class="op">=</span> <span class="va">resDeconvTCGA</span><span class="op">[</span>, <span class="st">"Stromal cells"</span><span class="op">]</span>,</span>
<span>  Entero <span class="op">=</span> <span class="va">resDeconvTCGA</span><span class="op">[</span>, <span class="st">"Enterocytes"</span><span class="op">]</span>,</span>
<span>  TypeSample <span class="op">=</span> <span class="va">TCGA.colon.se</span><span class="op">@</span><span class="va">colData</span><span class="op">$</span><span class="va">Tumor_Type</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">TypeSample</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Metastatic"</span>, <span class="st">"Recurrent"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html" class="external-link">melt</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">TypeSample</span>, y <span class="op">=</span> <span class="va">value</span>, fill <span class="op">=</span> <span class="va">TypeSample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">variable</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Estimated proportion"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">digitalDLSorteR</span><span class="fu">:::</span><span class="fu">default.colors</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Estimated proportions in TCGA data"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>      plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span>, hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>      legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Using Sample, TypeSample as id variables</span></span></code></pre>
<p><img src="boxplotResults_2_realModelWorkflow-1.png" width="1400" style="display: block; margin: auto;">
As it can be seen, <strong>digitalDLSorteR</strong> correctly estimates
the absence of tumor cells (<code>CRC</code>) in healthy samples. On the
other hand, the predicted proportion of enterocytes, epithelial and
stromal cells decrease in the tumoral samples, which makes sense
considering the infiltration of immune cells and the increased presence
of tumoral cells.</p>
</div>
<div class="section level2">
<h2 id="interpreting-the-neural-network-model">Interpreting the neural network model<a class="anchor" aria-label="anchor" href="#interpreting-the-neural-network-model"></a>
</h2>
<p>Finally, we have implemented a way to make the predictions made by
<strong>digitalDLSorteR</strong> more interpretable. This part was
developed for our new R package for deconvolution of spatial
transcriptomics data <a href="https://diegommcc.github.io/SpatialDDLS/index.html" class="external-link"><strong>SpatialDDLS</strong></a>,
and the methodology is explained in <span class="citation">Mañanes et
al. (2024)</span>.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DDLS.colon</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/interGradientsDL.html">interGradientsDL</a></span><span class="op">(</span><span class="va">DDLS.colon</span><span class="op">)</span></span></code></pre></div>
<p>We can explore the top 5 genes with the highest gradient for each
cell type to check which genes are being more used by the model:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top.gradients</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/topGradientsCellType.html">topGradientsCellType</a></span><span class="op">(</span></span>
<span>  <span class="va">DDLS.colon</span>, method <span class="op">=</span> <span class="st">"class"</span>, top.n.genes <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span></span>
<span>  <span class="va">top.gradients</span>, \<span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">$</span><span class="va">Positive</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   Anti-inflammatory_MFs B cells CD4+ T cells CD8+ T cells   ECs ECs_tumor</span></span>
<span><span class="co">## 1                 LYVE1    IGHD        KLRB1         GZMK FABP5      INSR</span></span>
<span><span class="co">## 2                  PLTP    CD19        ANXA1         GZMH CD320      AQP1</span></span>
<span><span class="co">## 3                  DAB2   FCRL1       GPR183        VCAM1  CD36     HSPG2</span></span>
<span><span class="co">## 4                LILRB5    SPIB          LTB        ITM2C SOCS3     SPRY1</span></span>
<span><span class="co">## 5                 SEPP1    PAX5          MAL         CCL5 HLA-E    FKBP1A</span></span>
<span><span class="co">##   Enterocytes Epithelial cells Epithelial_cancer_cells      MFs_SPP1+</span></span>
<span><span class="co">## 1        CMC1             ZG16                   ALDOB           CTSD</span></span>
<span><span class="co">## 2     UGT2B17             AQP8                  PTPN13          APOC1</span></span>
<span><span class="co">## 3       FCGRT             URAD                  GRIN2B RP11-1008C21.1</span></span>
<span><span class="co">## 4        SPIB             TFF3                    PLTP           CAPG</span></span>
<span><span class="co">## 5        CD63         HEXA-AS1                    CD63          MMP14</span></span>
<span><span class="co">##   Mast cells Myofibroblasts NK cells Pericytes Plasma_cells</span></span>
<span><span class="co">## 1       CTSG  RP11-400N13.3   FGFBP2     KCNJ8    LINC00152</span></span>
<span><span class="co">## 2      ANXA1          IGFL2     CMC1      CYBA         RGS1</span></span>
<span><span class="co">## 3      LTC4S          MMP14     GNLY      CD36        IGHA1</span></span>
<span><span class="co">## 4      GATA2          CCL11     GZMH     CNTN4         SRGN</span></span>
<span><span class="co">## 5       CMA1         IGFBP3    KRT86 LINC00152         GLDC</span></span>
<span><span class="co">##   Pro-inflammatory_MFs Regulatory T cells Smooth muscle cells Stromal cells</span></span>
<span><span class="co">## 1                 FCN1               BATF               RERGL            C7</span></span>
<span><span class="co">## 2              S100A12              RTKN2               CASQ2        FKBP11</span></span>
<span><span class="co">## 3                TIMP1            TNFRSF4               SUSD5          PID1</span></span>
<span><span class="co">## 4             SERPINB2               IL32                 DES        SEMA3E</span></span>
<span><span class="co">## 5                 VCAN                LTB               RBM24         ZFP36</span></span>
<span><span class="co">##   T follicular helper cells      cDC gamma delta T cells</span></span>
<span><span class="co">## 1                    CXCL13 HLA-DQA2               TRGC2</span></span>
<span><span class="co">## 2                     NR3C1   INSIG1               KLRC2</span></span>
<span><span class="co">## 3                       NMB     CD1C                TRDC</span></span>
<span><span class="co">## 4                    PTPN13  CLEC10A                CCL5</span></span>
<span><span class="co">## 5                      SRGN     PPA1                GZMA</span></span></code></pre>
<p>In addition, <strong>digitalDLSorteR</strong> also implements a
function to plot the top gradients per cell type as a heatmap:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hh</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotHeatmapGradsAgg.html">plotHeatmapGradsAgg</a></span><span class="op">(</span><span class="va">DDLS.colon</span>, top.n.genes <span class="op">=</span> <span class="fl">4</span>, method <span class="op">=</span> <span class="st">"class"</span><span class="op">)</span></span>
<span><span class="va">hh</span><span class="op">$</span><span class="va">Absolute</span></span></code></pre></div>
<p><img src="heatmapGradients_realModelWorkflow-1.png" width="2000" style="display: block; margin: auto;"></p>
<p>It is important to note that these markers should not be interpreted
as cell type markers. Rather, they serve as indications to help
interpret the model’s performance. In addition, due to the multivariate
nature of this approach, gradients are surrogates at the feature level
for predictions made considering all input variables collectively, and
thus caution should be exercised in drawing direct conclusions about
specific gene-cell type relationships.</p>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Ciriello2015" class="csl-entry">
Ciriello, G., M. L. Gatza, A. H. Beck, M. D. Wilkerson, S. K. Rhie, A.
Pastore, H. Zhang, et al. 2015. <span>“<span class="nocase"><span>C</span>omprehensive molecular portraits of
invasive lobular breast cancer</span>.”</span> <em>Cell</em> 163 (2):
506–19.
</div>
<div id="ref-Koboldt2012" class="csl-entry">
Koboldt, D. C., R. S. Fulton, M. D. McLellan, H. Schmidt, J.
Kalicki-Veizer, J. F. McMichael, L. L. Fulton, et al. 2012. <span>“<span class="nocase"><span>C</span>omprehensive molecular portraits of human
breast tumours</span>.”</span> <em>Nature</em> 490 (7418): 61–70.
</div>
<div id="ref-Lee2020" class="csl-entry">
Lee, H. O., Y. Hong, H. E. Etlioglu, Y. B. Cho, V. Pomella, B. Van den
Bosch, J. Vanhecke, et al. 2020. <span>“<span class="nocase"><span>L</span>ineage-dependent gene expression programs
influence the immune landscape of colorectal cancer</span>.”</span>
<em>Nat Genet</em> 52 (6): 594–603.
</div>
<div id="ref-Mananes2024" class="csl-entry">
Mañanes, D., I. Rivero-García, C. Relaño, Torres. M., D. Sancho, D.
Jimenez-Carretero, C. Torroja, and F. Sánchez-Cabo. 2024. <span>“<span class="nocase"><span>S</span>patialDDLS: an <span>R</span> package to
deconvolute spatial transcriptomics data using neural
networks</span>.”</span> <em>Bioinformatics</em> 40 (2).
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/diegommcc" class="external-link">Diego Mañanes</a>, <a href="https://www.cnic.es/en/carlos-torroja" class="external-link">Carlos Torroja</a>, <a href="https://www.cnic.es/en/fatima-sanchez-cabo" class="external-link">Fatima Sanchez-Cabo</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
