#' @importFrom methods setClass setOldClass setClassUnion
#' @importFrom utils packageVersion
#' @import SingleCellExperiment SummarizedExperiment
#' @importClassesFrom splatter ZINBParams
#' @importClassesFrom Matrix dgCMatrix
NULL

setOldClass(Classes = 'package_version')
setOldClass("keras.engine.sequential.Sequential")
setOldClass("keras_training_history")

setClassUnion("MatrixOrNULL", c("matrix", "NULL"))
setClassUnion("ListOrNULL", c("list", "NULL"))
setClassUnion("CharacterOrNULL", c("character", "NULL"))
setClassUnion("SingleCellExperimentOrNULL", c("SingleCellExperiment", "NULL"))
setClassUnion("ZINBParamsOrNULL", c("ZINBParams", "NULL"))
setClassUnion("KerasOrList", c("keras.engine.sequential.Sequential", "list"))

## ProbMatrixCellTypes class ----------------------------------------------------

#' The Class ProbMatrixCellTypes.
#'
#' The ProbMatrixCellTypes class is a data storage class that stores information
#' concerning to cell composition matrix used for the simulation of bulk
#' samples. This matrix corresponds with \code{prob.matrix} slot. The rest of
#' slots are additional information generated during the process and required
#' for subsequent steps.
#'
#' As described in Torroja and Sanchez-Cabo, 2019, the proportions are built by
#' five different methods in order to avoid biases due to the composition of the
#' bulk samples. In \code{plots} slot, different representations of these
#' probabilities are stored with the aim of offering a method to monitor the
#' different sets of samples generated during the process. These plots can be
#' displayed with \code{\link{showProbPlot}} function. See documentation for
#' details.
#'
#' @slot prob.matrix Matrix of cell proportions generated for the simulation of
#'   bulk samples. Rows correspond with bulk samples which will be generated
#'   (\eqn{i}), columns are the cell types present in single-cell data provided
#'   (\eqn{j}) and each entry is the proportion of \eqn{j} cell type in \eqn{i}
#'   sample.
#' @slot cell.names Matrix in which names of cells that will compound each bulk
#'   samples are stored.
#' @slot set.list List of cells ordered according to the cell type to which they
#'   belong.
#' @slot set Vector with the names of cells present in the object.
#' @slot exclusive.types Optional slot that contains the exclusive cell types on
#'   the experiment if they are provided. NULL by default.
#' @slot plots List of lists with plots showing the distribution of cell
#'   proportions generated by each method during the process.
#' @slot type.data Character with the type of data contained: training or test.
#'
#' @references Torroja, C. y SÃ¡nchez-Cabo, F. (2019). digitalDLSorter: A Deep
#' Learning algorithm to quantify immune cell populations based on scRNA-Seq
#' data. Frontiers in Genetics 10, 978. doi: \url{10.3389/fgene.2019.00978}
#'
#' @export ProbMatrixCellTypes
#'
ProbMatrixCellTypes <- setClass(
  Class = "ProbMatrixCellTypes",
  slots = c(
    prob.matrix = "MatrixOrNULL",
    cell.names = "MatrixOrNULL",
    set.list = "ListOrNULL",
    set = "CharacterOrNULL",
    exclusive.types = "CharacterOrNULL",
    plots = "ListOrNULL",
    type.data = "CharacterOrNULL"
  )
)

setMethod(
  f = "initialize", signature = "ProbMatrixCellTypes",
  definition = function(
    .Object,
    prob.matrix = NULL,
    cell.names = NULL,
    set.list = NULL,
    set = NULL,
    exclusive.types = NULL,
    plots = NULL,
    type.data = NULL
  ) {
    .Object@prob.matrix <- prob.matrix
    .Object@cell.names <- cell.names
    .Object@set.list <- set.list
    .Object@set <- set
    .Object@exclusive.types = exclusive.types
    .Object@plots <- plots
    .Object@type.data <- type.data
    return(.Object)
  }
)


setValidity(Class = "ProbMatrixCellTypes",
            method = function(object) {
              if (all(object@type.data != c("train", "test"))) {
                return(FALSE)
              } else {
                return(TRUE)
              }
            })

setMethod(f = "show",
          signature = "ProbMatrixCellTypes",
          definition = function(object) {
            # cat("An object of class", class(object), "\n")
            if (is.null(object@prob.matrix)) {
              cat("ProbMatrixCellTypes object empty")
            } else {
              cat(paste0("  Probability matrix for ",
                         object@type.data, ": "))
              cat(paste(dim(object@prob.matrix), c("bulk samples and",
                                                   "cell types"),
                        collapse = " "))
              if (!is.null(object@exclusive.types)) {
                cat("\n    Exclusive types:",
                    paste(object@exclusive.types, collapse = ", "))
              }
            }
          })


## DigitalDLSorterDNN class ----------------------------------------------------

#' The DigitalDLSorterDNN Class.
#'
#' The DigitalDLSorterDNN object stores the trained Deep Neural Network, the
#' training history of selected metrics and the results of prediction on test
#' data. After executing \code{\link{calculateEvalMetrics}}, it is alto possible
#' to find the results of the model evaluation.
#'
#' The steps related with Deep Learning are carried out with \code{keras}
#' package, so the model are stored in a R6 class, system used by the package.
#' If you want to save the object in an rds file, \code{digitalDLSorteR}
#' provides an \code{saveRDS} generic that transforms the keras model into a
#' native valid R object. Specifically, the model is converted into a list with
#' the architecture of the network and the weights learned during the training.
#' The is the minimum information to use the model as predictor. If you want to
#' maintain the optimizer state, see \code{\link{saveTrainedModelAsH5}}
#' function. If you want to store an object as rda file, see
#' \code{\link{preparingToSave}} function.
#'
#' @slot model Trained Deep Neural Network model. This slot can contain a R6
#'   \code{keras.engine.sequential.Sequential} object or a list with two
#'   elements: the architecture of the model and the resulting weights after
#'   training.
#' @slot training.history List with the evolution of the selected metrics during
#'   training.
#' @slot test.metrics Performance of the model on test data.
#' @slot test.pred Deconvolution results matrix of test data. Columns are
#'   cell types, rows are samples and each entry is the proportion of this cell
#'   type on this sample.
#' @slot cell.types Vector with the cell types to deconvolute.
#' @slot features Vector with features used during training. These features will
#'   be used for the following predictions.
#' @slot test.deconv.metrics Performance of the model on each sample of test data
#'   in comparison with the known cell proportions.
#'
#' @export DigitalDLSorterDNN
#'
DigitalDLSorterDNN <- setClass(
  Class = "DigitalDLSorterDNN",
  slots = c(
    model = "KerasOrList",
    training.history = "keras_training_history",
    test.metrics = "ListOrNULL",
    test.pred = "MatrixOrNULL",
    cell.types = "character",
    features = "character",
    test.deconv.metrics = "ListOrNULL"
  )
)

setMethod(
  f = "initialize", signature = "DigitalDLSorterDNN",
  definition = function(
    .Object,
    model = NULL,
    training.history = NULL,
    test.metrics = NULL,
    test.pred = NULL,
    cell.types = NULL,
    features = NULL,
    test.deconv.metrics = NULL
  ) {
    .Object@model <- model
    .Object@training.history <- training.history
    .Object@test.metrics <- test.metrics
    .Object@test.pred <- test.pred
    .Object@cell.types <- cell.types
    .Object@features <- features
    .Object@test.deconv.metrics <- test.deconv.metrics
    return(.Object)
  }
)

setMethod(
  f = "show",
  signature = "DigitalDLSorterDNN",
  definition = function(object) {
    # cat("An object of class", class(object), "\n")
    if (is.null(object@model)) {
      cat("DigitalDLSorterDNN object empty")
    } else {
      cat(paste("Trained model:", object@training.history$params$epochs,
                "epochs\n"))
      train.metrics <- lapply(object@training.history$metrics,
                              function(x) x[length(x)])
      cat("  Training metrics (last epoch):\n")
      cat(paste0("    ", names(train.metrics), ": ",
                 lapply(train.metrics, round, 4),
                 collapse = "\n"))
      cat("\n  Evaluation metrics on test data:\n")
      cat(paste0("    ", names(object@test.metrics), ": ",
                 lapply(object@test.metrics, round, 4),
                 collapse = "\n"))
      if (!is.null(test.deconv.metrics(object)) ||
          length(test.deconv.metrics(object)) > 0) {
        cat("\n  Performance evaluation over each sample: ")
        cat(paste(names(test.deconv.metrics(object)[[2]]), collapse = " "))
      }
    }
  }
)

setClassUnion("DigitalDLSorterDNNOrNULL", c("DigitalDLSorterDNN", "NULL"))


## DigitalDLSorter class -------------------------------------------------------

#' The DigitalDLSorter Class.
#'
#' The DigitalDLSorter object is the core of digitalDLSorteR. This object stores
#' the different intermediate data resulting from running pipeline from real
#' single-cell data to the trained Deep Neural Network, including the data on
#' which to carry out the process of devonvolution. Only it is used in the case
#' of building new deconvolution models. For deconvoluting bulk samples using
#' pre-trained models, see \code{\link{deconvDigitalDLSorter}} function.
#'
#' This object uses other classes to store the different type of data produced
#' during the process: \itemize{ \item \code{SingleCellExperiment} class for
#' single-cell RNA-seq data, using sparse matrix from the \code{Matrix} package
#' (\code{dgCMatrix} class) to store the matrix of counts. \item
#' \code{ZinbParams} class with the estimated parameters for the simulation of
#' new single-cell profiles. \item \code{SummarizedExperiment} class for storing
#' bulk RNA-seq data. In this case, it is possible to load all data in memory or
#' the use of HDF5 files as back-end by \code{DelayedArray} and \code{HDF5Array}
#' packages. See \code{\link{generateBulkSamples}} for details. \item
#' \code{\link{ProbMatrixCellTypes}} class for the composition cell matrices
#' built during the process. See \code{?ProbMatrixCellTypes} for details. \item
#' \code{\link{DigitalDLSorterDNN}} class for storing the trained Deep Neural
#' Network. This step is performed by \code{keras}. See
#' \code{\link{DigitalDLSorterDNN}} for details. }
#'
#' @slot single.cell.real Real single-cell data stored in a
#'   \code{SingleCellExperiment} object. The counts matrix is stored as a
#'   \code{dgCMatrix} object to optimize the amount of used memory.
#' @slot zinb.params \code{ZinbParams} object with estimated parameters for the
#'   simulation of new single-cell expression profiles.
#' @slot single.cell.simul Simulated single-cell expression profiles used for
#'   simulating bulk RNA-seq profiles with known cell composition.
#' @slot prob.cell.types \code{ProbMatrixCellTypes} class with the cell
#'   composition  matrix built for the simulation of bulk RNA-seq profiles. The
#'   entries determine the proportion of single-cell types that will constitute
#'   the simulated bulk samples.
#' @slot bulk.simul A list with two elements: train and test simulated bulk
#'   RNA-seq. This data are stored as a \code{SummarizedExperiment} object. We
#'   recommend the use of HDF5 file as a back-end due to the large amount of
#'   memory that they occupy.
#' @slot trained.model \code{\link{DigitalDLSorterDNN}} object with the trained
#'   model, different metrics obtained during the training and evaluation
#'   metrics from the application of the model on test data. After executing
#'   \code{\link{calculateEvalMetrics}}, it is alto possible to find the results
#'   of the model evaluation.
#' @slot deconv.data Optional slot where is possible to store new bulk samples
#'   for its deconvolution. It is a list whose name is the name of the data
#'   provided. It is possible to store more than one dataset to make
#'   predictions. See \code{\link{deconvDigitalDLSorterObj}} for details.
#' @slot deconv.results Slot where the results from the deconvolution process
#'   over \code{\link{deconv.data}} data are stored. It is a list whose name is
#'   the name of the data from which they come.
#' @slot project Name of the project.
#' @slot version Version of DigitalDLSorteR this object was built under.
#'
#'   The package can be used in two ways: to build new models of deconvolution
#'   from scRNA-seq data or to deconvolute bulk RNA-seq samples using
#'   pre-trtained models integrated into the package. If you want to build new
#'   models, see \code{\link{loadRealSCProfiles}} or
#'   \code{\link{loadFinalSCProfiles}} functions. If yoy want to use pre-trained
#'   models, see \code{\link{deconvDigitalDLSorter}} function.
#'
#' @exportClass DigitalDLSorter
#' @export DigitalDLSorter
#'
DigitalDLSorter <- setClass(
  Class = "DigitalDLSorter",
  slots = c(
    single.cell.real = "SingleCellExperimentOrNULL",
    zinb.params = "ZINBParamsOrNULL",
    single.cell.simul = "SingleCellExperimentOrNULL",
    prob.cell.types = "ListOrNULL",
    bulk.simul = "ListOrNULL",
    trained.model = "DigitalDLSorterDNNOrNULL",
    deconv.data = "ListOrNULL",
    deconv.results = "ListOrNULL",
    project = "character",
    version = "package_version"
  )
)


setMethod(
  f = "initialize", signature = "DigitalDLSorter",
  definition = function(
    .Object,
    single.cell.real = NULL,
    zinb.params = NULL,
    single.cell.simul = NULL,
    prob.cell.types = NULL,
    bulk.simul = NULL,
    trained.model = NULL,
    deconv.data = NULL,
    deconv.results = NULL,
    project = "DigitalDLSorterProject",
    version = packageVersion(pkg = "digitalDLSorteR")
  ) {
    .Object@single.cell.real <- single.cell.real
    .Object@zinb.params <- zinb.params
    .Object@single.cell.simul <- single.cell.simul
    .Object@prob.cell.types <- prob.cell.types
    .Object@bulk.simul <- bulk.simul
    .Object@trained.model <- trained.model
    .Object@deconv.data <- deconv.data
    .Object@deconv.results <- deconv.results
    .Object@project <- project
    .Object@version <- version
    return(.Object)
  }
)


.sceShow <- function(sce) {
  cat(" ", dim(sce)[1], "features and", dim(sce)[2], "cells\n")
  if (is.null(rownames(sce))) rownames.sce <- "---"
  else rownames.sce <- S4Vectors:::selectSome(rownames(sce), 6)
  if (identical(colnames(sce), character(0))) colnames.sce <- "---"
  else colnames.sce <- S4Vectors:::selectSome(colnames(sce), 6)
  cat("  rownames:", rownames.sce, "\n")
  cat("  colnames:", colnames.sce, "\n")
}

.bulkShow <- function(se) {
  cat("   ", dim(se)[1], "features and", dim(se)[2], "samples\n")
  if (is.null(rowData(se)[[1]])) rownames.se <- "---"
  else rownames.se <- S4Vectors:::selectSome(rowData(se)[[1]], 6)
  if (identical(colnames(se), character(0))) colnames.se <- "---"
  else colnames.se <- S4Vectors:::selectSome(colData(se)[[1]], 6)
  cat("    rownames:", rownames.se, "\n")
  cat("    colnames:", colnames.se, "\n")
}

.finalShow <- function(se) {
  cat("   ", dim(se)[2], "features and", dim(se)[1], "samples: ")
  n.bulk <- sum(grepl("Bulk\\.*", rowData(se)[[1]]))
  n.sc <- abs(n.bulk - dim(se)[1])
  cat(n.bulk, "bulk profiles and", n.sc, "single-cell profiles\n")
  # if (is.null(rowData(se)[[1]])) rownames.se <- "---"
  # else rownames.se <- S4Vectors:::selectSome(rowData(se)[[1]], 6)
  # if (identical(colnames(se), character(0))) colnames.se <- "---"
  # else colnames.se <- S4Vectors:::selectSome(colData(se)[[1]], 6)
  # cat("    rownames:", rownames.se, "\n")
  # cat("    colnames:", colnames.se, "\n")
}


.zinbModelShow <- function(zinb.model) {
  cat(paste0("ZinbParams object:\n",
             "  ", zinbwave::nSamples(zinb.model), " samples; ",
             "  ", zinbwave::nFeatures(zinb.model), " genes.\n",
             "  ", NCOL(zinbwave::getX_mu(zinb.model)),
             " sample-level covariate(s) (mu); ",
             "  ", NCOL(zinbwave::getX_pi(zinb.model)),
             " sample-level covariate(s) (pi);\n",
             "  ", NCOL(zinbwave::getV_mu(zinb.model)),
             " gene-level covariate(s) (mu); ",
             "  ", NCOL(zinbwave::getV_pi(zinb.model)),
             " gene-level covariate(s) (pi);\n",
             "  ", zinbwave::nFactors(zinb.model), " latent factor(s).\n"))
}

.allSlotsNull <- function(object) {
  list.slots <- list(
    "single.cell.real", "zinb.params", "single.cell.simul", "prob.cell.types",
    "bulk.simul", "trained.model", "deconv.data", "deconv.results"
  )
  res <- all(unlist(lapply(list.slots, function(x) is.null(do.call("@", list(object, x))))))
  if (res) return(TRUE)
  else return(FALSE)
}


setMethod(f = "show",
          signature = "DigitalDLSorter",
          definition = function(object) {
            if (.allSlotsNull(object)) {
              cat("An empty object of class", class(object), "\n")
              opt <- options(show.error.messages = FALSE)
              on.exit(options(opt))
              stop()
            } else {
              cat("An object of class", class(object), "\n")
            }
            if (!is.null(object@single.cell.real)) {
              cat("Real single-cell profiles:\n")
              .sceShow(object@single.cell.real)
            } else {
              cat("Real single-cell profiles:\n")
              .sceShow(S4Vectors::DataFrame())
            }
            if (!is.null(object@zinb.params)) {
              .zinbModelShow(object@zinb.params@model)
            }
            if (!is.null(object@single.cell.simul)) {
              cat("Simulated single-cell profiles:\n")
              .sceShow(object@single.cell.simul)
            }
            if (!is.null(object@prob.cell.types)) {
              cat("Probability matrices:\n")
              lapply(X = c("train", "test"), FUN = function(x) {
                if (x %in% names(object@prob.cell.types)) {
                  cat(show(object@prob.cell.types[[x]]), "\n")
                }
              })
            }
            if (!is.null(object@bulk.simul)) {
              cat("Simulated bulk samples:\n")
              lapply(X = c("train", "test"), FUN = function(x) {
                if (x %in% names(object@bulk.simul)) {
                  cat(paste(" ", x, "bulk samples:\n"))
                  .bulkShow(object@bulk.simul[[x]])
                }
              })
            }
            if (!is.null(object@trained.model)) {
              cat(show(object@trained.model), "\n")
            }
            cat("Project:", object@project, "\n")
          })
